{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import urllib.parse as up\n",
    "import os\n",
    "import psycopg2\n",
    "import re\n",
    "from Bio import Entrez\n",
    "from collections import Counter"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ficheiro GB   # fazer dowload diretamente do site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_genbank_input = input(\"ID NCBI\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.efetch(db='nucleotide',retmode='txt',id=id_genbank_input, rettype = \"gb\") # dowload do ficheiro\n",
    "record = SeqIO.read(handle, \"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# file = \"/home/sa_bruno/Documentos/GitHub/Projeto_LB/ARL15/ARL15.gb\"                \n",
    "# # file = \"Teste.gb\"\n",
    "# record = SeqIO.read(file, \"genbank\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tirar todas as variáveis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genbank\n",
    "ID = record.name\n",
    "Organism = record.annotations[\"organism\"]\n",
    "Accession = record.annotations[\"accessions\"][0]\n",
    "Definition = record.description\n",
    "Keywords = \"\"\n",
    "\n",
    "for c in record.annotations[\"keywords\"]:\n",
    "    Keywords += c + \", \"\n",
    "Keywords = Keywords[0: (len(Keywords) -2)]\n",
    "\n",
    "Sequence = record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ID)\n",
    "print(Organism)\n",
    "print(Accession)\n",
    "print(Definition)\n",
    "print(Keywords)\n",
    "print(Sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locus \n",
    "Size = len(record.seq)\n",
    "Molecular_type = record.annotations[\"molecule_type\"] + \" \" + record.annotations[\"topology\"]\n",
    "Genbank_division = record.annotations[\"data_file_division\"] \n",
    "Modification_date = record.annotations[\"date\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Size)\n",
    "print(Molecular_type)\n",
    "print(Genbank_division)\n",
    "print(Modification_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_authors(a):\n",
    "    exis_5 = re.search(r\"authors:\\s+([\\w\\s,\\.-]+)\", a)          #dúvida (como é que o ciclo sabe se isto está vazio ou não)       \n",
    "    if exis_5:\n",
    "        Authors = []\n",
    "        Authors_work = []\n",
    "        y = re.search(r\"authors:\\s+([\\w\\s,\\.-]+)\", a)    # isto aqui é o remark\n",
    "        Authors_work.append(y.group(1))\n",
    "        # else:\n",
    "        #     Consortium.append(\"------\")\n",
    "        help_authors = [] \n",
    "        Authors = str(Authors_work[0]).split(\"., \")\n",
    "        for i, c in enumerate(Authors):\n",
    "            f = c.split(\" and \")\n",
    "            help_authors.append([i, f])\n",
    "        c = len(Authors)\n",
    "        i = 0\n",
    "        while c > 0:\n",
    "            if len(help_authors[i][1]) > 1:\n",
    "                Authors.remove(Authors[i])\n",
    "                for o in help_authors[i][1]:\n",
    "                    Authors.append(o)\n",
    "\n",
    "            c -= 1\n",
    "            i += 1\n",
    "    else:\n",
    "        Authors = []\n",
    "\n",
    "    return Authors\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# referencias  # faz parte do PubMed INfo\n",
    "Pubmed = []\n",
    "Title = []\n",
    "Journal = []\n",
    "Consortium = []\n",
    "Remark = []\n",
    "Authors = []\n",
    "b = \"\"\n",
    "pubmed_trash = record.annotations[\"references\"]    # resolver isto \n",
    "c = len(pubmed_trash)\n",
    "i = 0\n",
    "while c > 0:\n",
    "    a = str(pubmed_trash[i]).replace(\"\\n\", \"§\")   \n",
    "    \n",
    "    exist_3 = re.search(r\"pubmed id:\\s+(\\d+)\", a)\n",
    "    if exist_3:\n",
    "        m = re.search(r\"pubmed id:\\s+(\\d+)\", a)                 # COnfirmar PubMEd haver sempre/nao\n",
    "        Pubmed.append(m.group(1))\n",
    "    else:\n",
    "        Pubmed.append(\"------\")\n",
    "\n",
    "    \n",
    "    n = re.search(r\"title:\\s+([\\w\\s]+)\", a)\n",
    "    Title.append(n.group(1))\n",
    "\n",
    "    exist_2 = re.search(r\"journal:\\s+([\\w\\s.()--,]+)\", a)\n",
    "    if exist_2:\n",
    "        j = re.search(r\"journal:\\s+([\\w\\s.()--,]+)\", a)\n",
    "        Journal.append(j.group(1))\n",
    "    else:\n",
    "        Journal.append(\"------\")\n",
    "\n",
    "    exis = re.search(r\"consrtm:\\s+([\\w\\s]+)\", a)\n",
    "    if exis:\n",
    "        k = re.search(r\"consrtm:\\s+([\\w\\s]+)\", a)    # isto aqui é o remark\n",
    "        Consortium.append(k.group(1))\n",
    "    else:\n",
    "        Consortium.append(\"------\")         # ter cuidado com isto dps\n",
    "    \n",
    "\n",
    "    exis_1 = re.search(r\"pubmed id:\\s+\\d+§comment:\\s+([\\w\\s\\d])\", a)\n",
    "    if exis_1:\n",
    "        l = re.search(r\"pubmed id:\\s+\\d+§comment:\\s+([\\w\\d\\s:\\[;\\(\\)\\.,!?<>{}\\\"]+)\", a)\n",
    "        Remark.append(l.group(1))\n",
    "    else:\n",
    "        Remark.append(\"------\") \n",
    "\n",
    "    Authors.append(get_authors(a))\n",
    "\n",
    "\n",
    "    \n",
    "    c -= 1\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = record.annotations[\"references\"][0]  #para usar dps\n",
    "a.authors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Authors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['8743090']\n",
      "['AIDS Res. Hum. Retroviruses 12 (7), 641-643 (1996)']\n",
      "['------']\n",
      "['------']\n",
      "[['Voevodin,A', 'Crandall,K.A', 'Seth,P.', 'al Mufti,S.']]\n"
     ]
    }
   ],
   "source": [
    "print(Pubmed)\n",
    "print(Journal)\n",
    "print(Consortium)\n",
    "print(Remark)\n",
    "print(Authors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Abstract = []                        # mudar para cada referenias \n",
    "DOI = []\n",
    "Affiliation = []\n",
    "Entrez.email = 'sa.bruno.2001@gmail.com'\n",
    "# handle = Entrez.efetch(db='pubmed',retmode='xml',id=\"15496913\")\n",
    "handle = Entrez.efetch(db='pubmed',retmode='xml',id=\"8743090\")\n",
    "resultado = Entrez.read(handle)\n",
    "q = str(resultado[\"PubmedArticle\"][0])\n",
    "q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = re.search(r\"AbstractText.:\\s+\\['([\\w\\s,\\.\\(\\)%-:]+)\", q)\n",
    "Abstract.append(o.group(1))\n",
    "\n",
    "p = re.search(r\"'pubmed'}.,\\s+StringElement.'([\\d\\w\\./]+)\" , q)\n",
    "DOI.append(p.group(1))\n",
    "\n",
    "t = re.search(r\"'Affiliation':\\s+'([,\\s\\d\\w\\./]+)\" , q)\n",
    "print(t.group(1))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Title)\n",
    "print(Abstract)\n",
    "print(DOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sequen\n",
    "\n",
    "temp_l = []\n",
    "for c in range(len(record.features)):\n",
    "    temp_l.append(record.features[c].type)\n",
    "\n",
    "y = Counter(temp_l)\n",
    "Gene = y[\"gene\"]\n",
    "# Regulatory = \n",
    "#(ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA))      # acabar esta parte dps  easy\n",
    "# so contar apartir daqui \n",
    "  # Counter({'gene': 5125, 'CDS': 4997, 'tRNA': 105, 'rRNA': 22, 'source': 1, 'tmRNA': 1})\n",
    "CDS = y[\"CDS\"]\n",
    "Regulatory = y[\"regulatory\"]\n",
    "exons = y[\"exons\"]\n",
    "poli_A_site = y[\"poli_A-site\"]                    #  poli_A-site\n",
    "misc_feature = y[\"misc_feature\"]\n",
    "mRNA = y[\"mRNA\"]\n",
    "print(y)\n",
    "Location_span = str(record.features[0].location)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Gene)\n",
    "print(CDS)\n",
    "print(Regulatory)\n",
    "print(poli_A_site)\n",
    "print(misc_feature)\n",
    "print(mRNA)\n",
    "print(Location_span)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# usar para dps\n",
    "CDS_list = []\n",
    "\n",
    "ID_cds = []\n",
    "Translacion_seq = []\n",
    "Localization = []\n",
    "ID_protein = []\n",
    "Protein = []\n",
    "for c in range(len(record.features)):\n",
    "    if record.features[c].type == \"CDS\":\n",
    "        CDS_list.append(record.features[c])\n",
    "\n",
    "# Translacion_seq\n",
    "for c in range(len(CDS_list)):\n",
    "    ID_cds.append(CDS_list[c])\n",
    "    Translacion_seq.append(CDS_list[c].qualifiers[\"translation\"])\n",
    "    Localization.append(CDS_list[c].location)\n",
    "    ID_protein.append(CDS_list[c].qualifiers[\"protein_id\"])\n",
    "    try:\n",
    "        Protein.append(CDS_list[c].qualifiers[\"product\"])\n",
    "    except:\n",
    "        Protein.append(\"------\")\n",
    "    \n",
    "#(ID_cds, Translacion_seq, Localization, ID_protein, Protein)) \n",
    "#    (CDS_list[0]).qualifiers[\"gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in range(len(ID_cds)):\n",
    "    print(ID_cds[c])\n",
    "    print()\n",
    "    print(Translacion_seq[c])\n",
    "    print()\n",
    "    print(Localization[c])\n",
    "    print()\n",
    "    print(ID_protein[c])\n",
    "    print()\n",
    "    print(Protein[c])\n",
    "    print()\n",
    "    print()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ligar Base Dados\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'conn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mProgrammingError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb Cell 25\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     cur\u001b[39m.\u001b[39;49mexecute(\u001b[39m\"\"\"\u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m        INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m \u001b[39m        VALUES (\u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m)\u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39m        \u001b[39;49m\u001b[39m\"\"\"\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m         (ID, Organism, Accession, Definition, Keywords))\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39mexcept\u001b[39;00m psycopg2\u001b[39m.\u001b[39mError \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/cursor.py:551\u001b[0m, in \u001b[0;36mexecute\u001b[0;34m(self, operation, params, multi)\u001b[0m\n\u001b[1;32m    548\u001b[0m     \u001b[39mif\u001b[39;00m psub\u001b[39m.\u001b[39mremaining \u001b[39m!=\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    549\u001b[0m         \u001b[39mraise\u001b[39;00m ProgrammingError(\n\u001b[1;32m    550\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mNot all parameters were used in the SQL statement\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 551\u001b[0m         )\n\u001b[1;32m    552\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:490\u001b[0m, in \u001b[0;36mMySQLConnection.cmd_query\u001b[0;34m(self, query, raw, buffered, raw_as_string)\u001b[0m\n\u001b[1;32m    489\u001b[0m     query \u001b[39m=\u001b[39m query\u001b[39m.\u001b[39mencode(\u001b[39m'\u001b[39m\u001b[39mutf-8\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> 490\u001b[0m result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_handle_result(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_cmd(ServerCmd\u001b[39m.\u001b[39;49mQUERY, query))\n\u001b[1;32m    492\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_have_next_result:\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:395\u001b[0m, in \u001b[0;36mMySQLConnection._handle_result\u001b[0;34m(self, packet)\u001b[0m\n\u001b[1;32m    394\u001b[0m \u001b[39melif\u001b[39;00m packet[\u001b[39m4\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m255\u001b[39m:\n\u001b[0;32m--> 395\u001b[0m     \u001b[39mraise\u001b[39;00m errors\u001b[39m.\u001b[39mget_exception(packet)\n\u001b[1;32m    397\u001b[0m \u001b[39m# We have a text result set\u001b[39;00m\n",
      "\u001b[0;31mProgrammingError\u001b[0m: 1046 (3D000): No database selected",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb Cell 25\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mErro na escrita na base de dados: \u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mformat(e) )    \n\u001b[1;32m     <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X26sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m      conn\u001b[39m.\u001b[39mclose()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'conn' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "try:\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Organism, Accession, Definition, Keywords))\n",
    "        \n",
    "        \n",
    "except psycopg2.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "     conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector as SQLC\n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "   host =\"127.0.0.1\",\n",
    "   user =\"root\",\n",
    "   password =\"login123\",\n",
    "   database = \"mydb\",\n",
    "   auth_plugin=\"mysql_native_password\"\n",
    ")\n",
    "cur = DataBase.cursor()\n",
    "\n",
    "#jdbc:mysql://127.0.0.1:3306/?user=root\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "ename": "DataError",
     "evalue": "1292 (22007): Incorrect date value: '24-MAR-1997' for column 'Modification_date' at row 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mDataError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb Cell 27\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m cur\u001b[39m.\u001b[39;49mexecute(\u001b[39m\"\"\"\u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39m        INSERT INTO Locus (ID, Size, Molecular_type, Genbank_division, Modification_date)    \u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m        VALUES (\u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m, \u001b[39;49m\u001b[39m%s\u001b[39;49;00m\u001b[39m)\u001b[39;49m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m \u001b[39m        \u001b[39;49m\u001b[39m\"\"\"\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m         (ID, Size, Molecular_type, Genbank_division, Modification_date))\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m \u001b[39m#     cur.execute(\"\"\"\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39m#         INSERT INTO Sequences (ID, Sequence, Count_A, Count_C, Count_T, Count_G)    \u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m \u001b[39m#         VALUES (%s, %s, %s, %s, %s, %s,%s)\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=60'>61</a>\u001b[0m \u001b[39m#         \"\"\",\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=61'>62</a>\u001b[0m \u001b[39m#         (Authors, Affiliation))    \u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X30sZmlsZQ%3D%3D?line=62'>63</a>\u001b[0m cur\u001b[39m.\u001b[39mclose\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/cursor.py:551\u001b[0m, in \u001b[0;36mexecute\u001b[0;34m(self, operation, params, multi)\u001b[0m\n\u001b[1;32m    547\u001b[0m     stmt \u001b[39m=\u001b[39m RE_PY_PARAM\u001b[39m.\u001b[39msub(psub, stmt)\n\u001b[1;32m    548\u001b[0m     \u001b[39mif\u001b[39;00m psub\u001b[39m.\u001b[39mremaining \u001b[39m!=\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    549\u001b[0m         \u001b[39mraise\u001b[39;00m ProgrammingError(\n\u001b[1;32m    550\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mNot all parameters were used in the SQL statement\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 551\u001b[0m         )\n\u001b[1;32m    552\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    553\u001b[0m     \u001b[39mraise\u001b[39;00m ProgrammingError(\n\u001b[1;32m    554\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCould not process parameters: \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mtype\u001b[39m(params)\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m(\u001b[39m\u001b[39m{\u001b[39;00mparams\u001b[39m}\u001b[39;00m\u001b[39m),\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    555\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m it must be of type list, tuple or dict\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    556\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:490\u001b[0m, in \u001b[0;36mMySQLConnection.cmd_query\u001b[0;34m(self, query, raw, buffered, raw_as_string)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(query, \u001b[39mbytes\u001b[39m):\n\u001b[1;32m    489\u001b[0m     query \u001b[39m=\u001b[39m query\u001b[39m.\u001b[39mencode(\u001b[39m'\u001b[39m\u001b[39mutf-8\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> 490\u001b[0m result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_handle_result(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_cmd(ServerCmd\u001b[39m.\u001b[39;49mQUERY, query))\n\u001b[1;32m    492\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_have_next_result:\n\u001b[1;32m    493\u001b[0m     \u001b[39mraise\u001b[39;00m errors\u001b[39m.\u001b[39mInterfaceError(\n\u001b[1;32m    494\u001b[0m         \u001b[39m'\u001b[39m\u001b[39mUse cmd_query_iter for statements with multiple queries.\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:395\u001b[0m, in \u001b[0;36mMySQLConnection._handle_result\u001b[0;34m(self, packet)\u001b[0m\n\u001b[1;32m    393\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_handle_eof(packet)\n\u001b[1;32m    394\u001b[0m \u001b[39melif\u001b[39;00m packet[\u001b[39m4\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m255\u001b[39m:\n\u001b[0;32m--> 395\u001b[0m     \u001b[39mraise\u001b[39;00m errors\u001b[39m.\u001b[39mget_exception(packet)\n\u001b[1;32m    397\u001b[0m \u001b[39m# We have a text result set\u001b[39;00m\n\u001b[1;32m    398\u001b[0m column_count \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_protocol\u001b[39m.\u001b[39mparse_column_count(packet)\n",
      "\u001b[0;31mDataError\u001b[0m: 1292 (22007): Incorrect date value: '24-MAR-1997' for column 'Modification_date' at row 1"
     ]
    }
   ],
   "source": [
    "\n",
    "cur.execute(\"\"\"\n",
    "        INSERT INTO Locus (ID, Size, Molecular_type, Genbank_division, Modification_date)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Size, Molecular_type, Genbank_division, Modification_date))\n",
    "\n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO Sequences (ID, Sequence, Count_A, Count_C, Count_T, Count_G)    \n",
    "#         VALUES (%s, %s, %s, %s, %s, %s,%s)\n",
    "#         \"\"\",\n",
    "#         (ID, Sequence, Count_A, Count_C, Count_T, Count_G))   \n",
    "\n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO Pubmed_Info (ID_pubmed, Title, Abstract, Pubmed_Infocol, Authors)    \n",
    "#         VALUES (%s, %s, %s, %s, %s)\n",
    "#         \"\"\",\n",
    "#         (Pubmed, Title, Abstract, DOI, Authors))     # id_pubmed na db\n",
    "    \n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO References (ID, Pubmed, Title, Journal, Consortium, Remark)    \n",
    "#         VALUES (%s, %s, %s, %s, %s, %s))\n",
    "#         \"\"\",\n",
    "#         (ID, Pubmed, Journal, Consortium, Remark))   \n",
    "#     # \n",
    "# import psycopg2\n",
    "\n",
    "# conn = psycopg2.connect('postgresql://geobox:xxxxxxxxxxxx@geo.di.uminho.pt/bioinformatica')\n",
    "\n",
    "# conn.autocommit = True\n",
    "\n",
    "# cur = conn.cursor()\n",
    "# try:\n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \n",
    "#         VALUES (%s, %s, %s, %s, %s)\n",
    "#         \"\"\",\n",
    "#         (ID, Organism, Accession, Definition, Keywords))\n",
    "        \n",
    "        \n",
    "# except psycopg2.Error as e:\n",
    "\n",
    "#     print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "# finally:\n",
    "#     conn.close()\n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO Features (ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA, Location_span)    \n",
    "#         VALUES (%s, %s, %s, %s, %s, %s, %s, %s))\n",
    "#         \"\"\",\n",
    "#         (ID, CDS, Gene, Regulatory, exons, poli_A_site, misc_feature, mRNA, Location_span))   \n",
    "\n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO CDS (ID_CDS, Translacion_seq, Localization, ID_protein, Protein)    \n",
    "#         VALUES (%s, %s, %s, %s, %s))\n",
    "#         \"\"\",\n",
    "#         (ID, Translacion_seq, Localization, ID_protein, Protein)) \n",
    "\n",
    "    \n",
    "#     cur.execute(\"\"\"\n",
    "#         INSERT INTO Authors (ID_authors, Affiliation)    \n",
    "#         VALUES (%s, %s, %s))\n",
    "#         \"\"\",\n",
    "#         (Authors, Affiliation))    \n",
    "cur.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "ename": "ProgrammingError",
     "evalue": "1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\"qcratyzs\".\"Locus\" (\n  \"ID\" VARCHAR(50) NOT NULL,\n  \"Size\" INT NOT NULL,\n  \"Mole' at line 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mProgrammingError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb Cell 28\u001b[0m in \u001b[0;36m<cell line: 152>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m ç \u001b[39m=\u001b[39m (\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \n\u001b[1;32m      <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=149'>150</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=150'>151</a>\u001b[0m )\n\u001b[0;32m--> <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=151'>152</a>\u001b[0m cur\u001b[39m.\u001b[39;49mexecute(ç)\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/sa_bruno/Documentos/GitHub/Projeto_Algoritmos-Introducao/IAP/Final/Final.ipynb#X31sZmlsZQ%3D%3D?line=153'>154</a>\u001b[0m conn\u001b[39m.\u001b[39mclose()\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/cursor.py:551\u001b[0m, in \u001b[0;36mexecute\u001b[0;34m(self, operation, params, multi)\u001b[0m\n\u001b[1;32m    547\u001b[0m     stmt \u001b[39m=\u001b[39m RE_PY_PARAM\u001b[39m.\u001b[39msub(psub, stmt)\n\u001b[1;32m    548\u001b[0m     \u001b[39mif\u001b[39;00m psub\u001b[39m.\u001b[39mremaining \u001b[39m!=\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    549\u001b[0m         \u001b[39mraise\u001b[39;00m ProgrammingError(\n\u001b[1;32m    550\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mNot all parameters were used in the SQL statement\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 551\u001b[0m         )\n\u001b[1;32m    552\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    553\u001b[0m     \u001b[39mraise\u001b[39;00m ProgrammingError(\n\u001b[1;32m    554\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCould not process parameters: \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mtype\u001b[39m(params)\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m(\u001b[39m\u001b[39m{\u001b[39;00mparams\u001b[39m}\u001b[39;00m\u001b[39m),\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    555\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m it must be of type list, tuple or dict\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    556\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:490\u001b[0m, in \u001b[0;36mMySQLConnection.cmd_query\u001b[0;34m(self, query, raw, buffered, raw_as_string)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(query, \u001b[39mbytes\u001b[39m):\n\u001b[1;32m    489\u001b[0m     query \u001b[39m=\u001b[39m query\u001b[39m.\u001b[39mencode(\u001b[39m'\u001b[39m\u001b[39mutf-8\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> 490\u001b[0m result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_handle_result(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_send_cmd(ServerCmd\u001b[39m.\u001b[39;49mQUERY, query))\n\u001b[1;32m    492\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_have_next_result:\n\u001b[1;32m    493\u001b[0m     \u001b[39mraise\u001b[39;00m errors\u001b[39m.\u001b[39mInterfaceError(\n\u001b[1;32m    494\u001b[0m         \u001b[39m'\u001b[39m\u001b[39mUse cmd_query_iter for statements with multiple queries.\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/mestrado_final/lib/python3.9/site-packages/mysql/connector/connection.py:395\u001b[0m, in \u001b[0;36mMySQLConnection._handle_result\u001b[0;34m(self, packet)\u001b[0m\n\u001b[1;32m    393\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_handle_eof(packet)\n\u001b[1;32m    394\u001b[0m \u001b[39melif\u001b[39;00m packet[\u001b[39m4\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m255\u001b[39m:\n\u001b[0;32m--> 395\u001b[0m     \u001b[39mraise\u001b[39;00m errors\u001b[39m.\u001b[39mget_exception(packet)\n\u001b[1;32m    397\u001b[0m \u001b[39m# We have a text result set\u001b[39;00m\n\u001b[1;32m    398\u001b[0m column_count \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_protocol\u001b[39m.\u001b[39mparse_column_count(packet)\n",
      "\u001b[0;31mProgrammingError\u001b[0m: 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\"qcratyzs\".\"Locus\" (\n  \"ID\" VARCHAR(50) NOT NULL,\n  \"Size\" INT NOT NULL,\n  \"Mole' at line 1"
     ]
    }
   ],
   "source": [
    "ç = (\n",
    "\n",
    "\"\"\"\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Locus\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Size\" INT NOT NULL,\n",
    "  \"Molecular_type\" VARCHAR(50) NOT NULL,\n",
    "  \"Genbank_division\" VARCHAR(3) NOT NULL,\n",
    "  \"Modification_date\" DATE NOT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Sequences\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Sequences\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Sequence\" VARCHAR(2000) NOT NULL,\n",
    "  \"Length\" INT NULL,\n",
    "  \"Count_A\" INT NULL,\n",
    "  \"Count_C\" INT NULL,\n",
    "  \"Count_T\" INT NULL,\n",
    "  \"Count_G\" INT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"References\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"References\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Pubmed\" VARCHAR(50) NOT NULL,\n",
    "  \"Title\" VARCHAR(200) NULL,\n",
    "  \"Journal\" VARCHAR(200) NULL,\n",
    "  \"Consortium\" VARCHAR(200) NULL,\n",
    "  \"Remark\" VARCHAR(200) NULL,\n",
    "  PRIMARY KEY (\"ID\", \"Pubmed\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Source\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Source\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Location_span\" VARCHAR(50) NOT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"CDS\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"CDS\" (\n",
    "  \"ID_CDS\" VARCHAR(50) NOT NULL,\n",
    "  \"Translacion_seq\" VARCHAR(2000) NOT NULL,\n",
    "  \"Localization\" VARCHAR(20) NOT NULL,\n",
    "  \"ID_protein\" VARCHAR(50) NOT NULL,\n",
    "  \"Protein\" VARCHAR(50) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_CDS\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Features\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Features\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"CDS\" VARCHAR(50) NOT NULL,\n",
    "  \"Gene\" INT NULL,\n",
    "  \"Regulatory\" INT NULL,\n",
    "  \"exons\" INT NULL,\n",
    "  \"poli_A-site\" INT NULL,\n",
    "  \"misc_feature\" INT NULL,\n",
    "  \"mRNA\" INT NULL,\n",
    "  PRIMARY KEY (\"ID\"),\n",
    "  CONSTRAINT \"Features-Source\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Source\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"features-CDS\"\n",
    "    FOREIGN KEY (\"CDS\")\n",
    "    REFERENCES \"qcratyzs\".\"CDS\" (\"ID_CDS\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Genbank\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Genbank\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Organism\" VARCHAR(200) NOT NULL,\n",
    "  \"Accession\" VARCHAR(50) NOT NULL,\n",
    "  \"Definition\" VARCHAR(200) NULL,\n",
    "  \"Keywords\" VARCHAR(200) NULL,\n",
    "  PRIMARY KEY (\"ID\"),\n",
    "  CONSTRAINT \"Genbank-locus\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Locus\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-sequence\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Sequences\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-references\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"References\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-features\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Features\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Pubmed_Info\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Pubmed_Info\" (\n",
    "  \"ID_pubmed\" VARCHAR(50) NOT NULL,\n",
    "  \"Title\" VARCHAR(200) NOT NULL,\n",
    "  \"Abstract\" VARCHAR(2000) NOT NULL,\n",
    "  \"Pubmed_Infocol\" VARCHAR(50) NOT NULL,\n",
    "  \"Authors\" VARCHAR(100) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_pubmed\", \"Pubmed_Infocol\", \"Authors\"),\n",
    "  CONSTRAINT \"Pubmed-references\"\n",
    "    FOREIGN KEY (\"ID_pubmed\")\n",
    "    REFERENCES \"qcratyzs\".\"References\" (\"Pubmed\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Authors\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Authors\" (\n",
    "  \"ID_authors\" VARCHAR(100) NOT NULL,\n",
    "  \"Name\" VARCHAR(200) NOT NULL,\n",
    "  \"Affiliation\" VARCHAR(200) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_authors\"),\n",
    "  CONSTRAINT \"Autores- pubmed\"\n",
    "    FOREIGN KEY (\"ID_authors\")\n",
    "    REFERENCES \"qcratyzs\".\"Pubmed_Info\" (\"Authors\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "\"\"\"\n",
    ")\n",
    "cur.execute(ç)\n",
    "\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mestrado_final",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f2307cf6e279acbb4fe5a904c1ba66075afa04f4e5f00d0d9abed1d852b5aca9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
