{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import urllib.parse as up\n",
    "import os\n",
    "import psycopg2\n",
    "import re\n",
    "from Bio import Entrez\n",
    "from collections import Counter"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ficheiro GB   # fazer dowload diretamente do site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# file = \"/home/sa_bruno/Documentos/GitHub/Projeto_LB/ARL15/ARL15.gb\"                \n",
    "file = \"Teste.gb\"\n",
    "record = SeqIO.read(file, \"genbank\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tirar todas as variáveis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genbank\n",
    "ID = record.name\n",
    "Organism = record.annotations[\"organism\"]\n",
    "Accession = record.annotations[\"accessions\"][0]\n",
    "Definition = record.description\n",
    "Keywords = \"\"\n",
    "\n",
    "for c in record.annotations[\"keywords\"]:\n",
    "    Keywords += c + \", \"\n",
    "Keywords = Keywords[0: (len(Keywords) -2)]\n",
    "\n",
    "Sequence = record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ID)\n",
    "print(Organism)\n",
    "print(Accession)\n",
    "print(Definition)\n",
    "print(Keywords)\n",
    "print(Sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Locus \n",
    "Size = len(record.seq)\n",
    "Molecular_type = record.annotations[\"molecule_type\"] + \" \" + record.annotations[\"topology\"]\n",
    "Genbank_division = record.annotations[\"data_file_division\"] \n",
    "Modification_date = record.annotations[\"date\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Size)\n",
    "print(Molecular_type)\n",
    "print(Genbank_division)\n",
    "print(Modification_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# referencias  # faz parte do PubMed INfo\n",
    "Pubmed = []\n",
    "Title = []\n",
    "Journal = []\n",
    "Consortium = []\n",
    "Remark = []\n",
    "Authors = []\n",
    "b = \"\"\n",
    "pubmed_trash = record.annotations[\"references\"]    # resolver isto \n",
    "c = len(pubmed_trash)\n",
    "i = 0\n",
    "while c > 0:\n",
    "    a = str(pubmed_trash[i]).replace(\"\\n\", \"§\")   \n",
    "    \n",
    "    exist_3 = re.search(r\"pubmed id:\\s+(\\d+)\", a)\n",
    "    if exist_3:\n",
    "        m = re.search(r\"pubmed id:\\s+(\\d+)\", a)                 # COnfirmar PubMEd haver sempre/nao\n",
    "        Pubmed.append(m.group(1))\n",
    "\n",
    "    \n",
    "    n = re.search(r\"title:\\s+([\\w\\s]+)\", a)\n",
    "    Title.append(n.group(1))\n",
    "\n",
    "    exist_2 = re.search(r\"journal:\\s+([\\w\\s.()--,]+)\", a)\n",
    "    if exist_2:\n",
    "        j = re.search(r\"journal:\\s+([\\w\\s.()--,]+)\", a)\n",
    "        Journal.append(j.group(1))\n",
    "    else:\n",
    "        Journal.append(\"------\")\n",
    "\n",
    "    exis = re.search(r\"consrtm:\\s+([\\w\\s]+)\", a)\n",
    "    if exis:\n",
    "        k = re.search(r\"consrtm:\\s+([\\w\\s]+)\", a)    # isto aqui é o remark\n",
    "        Consortium.append(k.group(1))\n",
    "    else:\n",
    "        Consortium.append(\"------\")         # ter cuidado com isto dps\n",
    "    \n",
    "\n",
    "    exis_1 = re.search(r\"pubmed id:\\s+\\d+§comment:\\s+([\\w\\s\\d])\", a)\n",
    "    if exis_1:\n",
    "        l = re.search(r\"pubmed id:\\s+\\d+§comment:\\s+([\\w\\d\\s:\\[;\\(\\)\\.,!?<>{}\\\"]+)\", a)\n",
    "        Remark.append(l.group(1))\n",
    "    else:\n",
    "        Remark.append(\"------\") \n",
    "\n",
    "    Authors.append(get_authors(a))\n",
    "\n",
    "\n",
    "    \n",
    "    c -= 1\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_authors(a):\n",
    "    exis_5 = re.search(r\"authors:\\s+([\\w\\s,\\.-]+)\", a)\n",
    "    if exis_5:\n",
    "        Authors = []\n",
    "        Authors_work = []\n",
    "        y = re.search(r\"authors:\\s+([\\w\\s,\\.-]+)\", a)    # isto aqui é o remark\n",
    "        Authors_work.append(y.group(1))\n",
    "        # else:\n",
    "        #     Consortium.append(\"------\")\n",
    "        help_authors = [] \n",
    "        Authors = str(Authors_work[0]).split(\"., \")\n",
    "        for i, c in enumerate(Authors):\n",
    "            f = c.split(\" and \")\n",
    "            help_authors.append([i, f])\n",
    "        c = len(Authors)\n",
    "        i = 0\n",
    "        while c > 0:\n",
    "            if len(help_authors[i][1]) > 1:\n",
    "                Authors.remove(Authors[i])\n",
    "                for o in help_authors[i][1]:\n",
    "                    Authors.append(o)\n",
    "\n",
    "            c -= 1\n",
    "            i += 1\n",
    "    else:\n",
    "        Authors = []\n",
    "\n",
    "        \n",
    "    return Authors\n",
    "get_authors(a)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Authors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Pubmed)\n",
    "print(Journal)\n",
    "print(Consortium)\n",
    "print(Remark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Abstract = []\n",
    "DOI = []\n",
    "Entrez.email = 'sa.bruno.2001@gmail.com'\n",
    "# handle = Entrez.efetch(db='pubmed',retmode='xml',id=\"15496913\")\n",
    "handle = Entrez.efetch(db='pubmed',retmode='xml',id=\"32678145\")\n",
    "resultado = Entrez.read(handle)\n",
    "q = str(resultado[\"PubmedArticle\"][0])\n",
    "q\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = re.search(r\"AbstractText.:\\s+\\['([\\w\\s,\\.\\(\\)%-:]+)\", q)\n",
    "Abstract.append(o.group(1))\n",
    "\n",
    "p = re.search(r\"'pubmed'}.,\\s+StringElement.'([\\d\\w\\./]+)\" , q)\n",
    "DOI.append(p.group(1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(Title)\n",
    "print(Abstract)\n",
    "print(DOI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'gene': 5125, 'CDS': 4997, 'tRNA': 105, 'rRNA': 22, 'source': 1, 'tmRNA': 1})\n"
     ]
    }
   ],
   "source": [
    "#sequen\n",
    "\n",
    "temp_l = []\n",
    "for c in range(len(record.features)):\n",
    "    temp_l.append(record.features[c].type)\n",
    "\n",
    "y = Counter(temp_l)\n",
    "Gene = y[\"gene\"]\n",
    "# Regulatory = \n",
    "print(y)\n",
    "#(ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA))      # acabar esta parte dps  easy\n",
    "    # so contar apartir daqui \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# usar para dps\n",
    "CDS = []\n",
    "for c in range(len(record.features)):\n",
    "    if record.features[c].type == \"CDS\":\n",
    "        CDS.append(record.features[c])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ligar Base Dados\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "\n",
    "\n",
    "conn = psycopg2.connect(\"postgres://qcratyzs:FELOcVIRwFjW6cw19YipUTM1n9VAc8dW@mel.db.elephantsql.com/qcratyzs\")\n",
    "conn.autocommit = True\n",
    "\n",
    "cur = conn.cursor()\n",
    "try:\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Organism, Accession, Definition, Keywords))\n",
    "        \n",
    "        \n",
    "except psycopg2.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "# finally:\n",
    "#     conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Locus (ID, Size, Molecular_type, Genbank_division, Modification_date)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Size, Molecular_type, Genbank_division, Modification_date))\n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Sequences (ID, Sequence, Count_A, Count_C, Count_T, Count_G)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s,%s)\n",
    "        \"\"\",\n",
    "        (ID, Sequence, Count_A, Count_C, Count_T, Count_G))   #retirei o length porque já tem no locus\n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Pubmed_Info (ID_pubmed, Title, Abstract, Pubmed_Infocol, Authors)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (Pubmed, Title, Abstract, DOI, Authors))     # id_pubmed na db\n",
    "    \n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO References (ID, Pubmed, Title, Journal, Consortium, Remark)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID, Pubmed, Journal, Consortium, Remark))   \n",
    "import psycopg2\n",
    "\n",
    "\n",
    "conn = psycopg2.connect('postgresql://geobox:xxxxxxxxxxxx@geo.di.uminho.pt/bioinformatica')\n",
    "conn.autocommit = True\n",
    "\n",
    "cur = conn.cursor()\n",
    "try:\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Organism, Accession, Definition, Keywords))\n",
    "        \n",
    "        \n",
    "except psycopg2.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    conn.close()\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Features (ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA))   \n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO CDS (ID_CDS, Translacion_seq, Localization, ID_protein, Protein)    \n",
    "        VALUES (%s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID, Translacion_seq, Localization, ID_protein, Protein)) \n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Source (ID, Location_span)    \n",
    "        VALUES (%s, %s))\n",
    "        \"\"\",\n",
    "        (ID, Location_span)) \n",
    "    \n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Authors (ID_authors, Affiliation)    \n",
    "        VALUES (%s, %s, %s))\n",
    "        \"\"\",\n",
    "        (Authors, Affiliation))    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cur.execute(\n",
    "\n",
    "\"\"\"\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Locus\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Size\" INT NOT NULL,\n",
    "  \"Molecular_type\" VARCHAR(50) NOT NULL,\n",
    "  \"Genbank_division\" VARCHAR(3) NOT NULL,\n",
    "  \"Modification_date\" DATE NOT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Sequences\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Sequences\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Sequence\" VARCHAR(2000) NOT NULL,\n",
    "  \"Length\" INT NULL,\n",
    "  \"Count_A\" INT NULL,\n",
    "  \"Count_C\" INT NULL,\n",
    "  \"Count_T\" INT NULL,\n",
    "  \"Count_G\" INT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"References\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"References\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Pubmed\" VARCHAR(50) NOT NULL,\n",
    "  \"Title\" VARCHAR(200) NULL,\n",
    "  \"Journal\" VARCHAR(200) NULL,\n",
    "  \"Consortium\" VARCHAR(200) NULL,\n",
    "  \"Remark\" VARCHAR(200) NULL,\n",
    "  PRIMARY KEY (\"ID\", \"Pubmed\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Source\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Source\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Location_span\" VARCHAR(50) NOT NULL,\n",
    "  PRIMARY KEY (\"ID\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"CDS\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"CDS\" (\n",
    "  \"ID_CDS\" VARCHAR(50) NOT NULL,\n",
    "  \"Translacion_seq\" VARCHAR(2000) NOT NULL,\n",
    "  \"Localization\" VARCHAR(20) NOT NULL,\n",
    "  \"ID_protein\" VARCHAR(50) NOT NULL,\n",
    "  \"Protein\" VARCHAR(50) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_CDS\"));\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Features\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Features\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"CDS\" VARCHAR(50) NOT NULL,\n",
    "  \"Gene\" INT NULL,\n",
    "  \"Regulatory\" INT NULL,\n",
    "  \"exons\" INT NULL,\n",
    "  \"poli_A-site\" INT NULL,\n",
    "  \"misc_feature\" INT NULL,\n",
    "  \"mRNA\" INT NULL,\n",
    "  PRIMARY KEY (\"ID\"),\n",
    "  CONSTRAINT \"Features-Source\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Source\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"features-CDS\"\n",
    "    FOREIGN KEY (\"CDS\")\n",
    "    REFERENCES \"qcratyzs\".\"CDS\" (\"ID_CDS\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Genbank\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Genbank\" (\n",
    "  \"ID\" VARCHAR(50) NOT NULL,\n",
    "  \"Organism\" VARCHAR(200) NOT NULL,\n",
    "  \"Accession\" VARCHAR(50) NOT NULL,\n",
    "  \"Definition\" VARCHAR(200) NULL,\n",
    "  \"Keywords\" VARCHAR(200) NULL,\n",
    "  PRIMARY KEY (\"ID\"),\n",
    "  CONSTRAINT \"Genbank-locus\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Locus\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-sequence\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Sequences\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-references\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"References\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION,\n",
    "  CONSTRAINT \"Genbank-features\"\n",
    "    FOREIGN KEY (\"ID\")\n",
    "    REFERENCES \"qcratyzs\".\"Features\" (\"ID\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Pubmed_Info\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Pubmed_Info\" (\n",
    "  \"ID_pubmed\" VARCHAR(50) NOT NULL,\n",
    "  \"Title\" VARCHAR(200) NOT NULL,\n",
    "  \"Abstract\" VARCHAR(2000) NOT NULL,\n",
    "  \"Pubmed_Infocol\" VARCHAR(50) NOT NULL,\n",
    "  \"Authors\" VARCHAR(100) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_pubmed\", \"Pubmed_Infocol\", \"Authors\"),\n",
    "  CONSTRAINT \"Pubmed-references\"\n",
    "    FOREIGN KEY (\"ID_pubmed\")\n",
    "    REFERENCES \"qcratyzs\".\"References\" (\"Pubmed\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "-- -----------------------------------------------------\n",
    "-- Table \"qcratyzs\".\"Authors\"\n",
    "-- -----------------------------------------------------\n",
    "CREATE TABLE IF NOT EXISTS \"qcratyzs\".\"Authors\" (\n",
    "  \"ID_authors\" VARCHAR(100) NOT NULL,\n",
    "  \"Name\" VARCHAR(200) NOT NULL,\n",
    "  \"Affiliation\" VARCHAR(200) NOT NULL,\n",
    "  PRIMARY KEY (\"ID_authors\"),\n",
    "  CONSTRAINT \"Autores- pubmed\"\n",
    "    FOREIGN KEY (\"ID_authors\")\n",
    "    REFERENCES \"qcratyzs\".\"Pubmed_Info\" (\"Authors\")\n",
    "    ON DELETE NO ACTION\n",
    "    ON UPDATE NO ACTION);\n",
    "\n",
    "\n",
    "\"\"\"\n",
    ")\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mestrado_final",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f2307cf6e279acbb4fe5a904c1ba66075afa04f4e5f00d0d9abed1d852b5aca9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
