{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import Entrez\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"L42022.1.gb\"                \n",
    "record = SeqIO.read(file, \"genbank\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: L42022.1\n",
      "Description: Human immunodeficiency virus type 1 (isolate genotype C, I5) gag gene, partial cds\n",
      "Name: HIVI5C\n",
      "Lenght:231\n",
      "Type: DNA linear\n",
      "Accessions['L42022']\n",
      "Source: Human immunodeficiency virus 1 (HIV-1)\n",
      "\n",
      "Publications:\n",
      "HIV type 1 subtypes B and C from new regions of India and Indian and Ethiopian expatriates in Kuwait\n",
      "AIDS Res. Hum. Retroviruses 12 (7), 641-643 (1996)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"ID: \" + record.id)\n",
    "print(\"Description: \" + record.description)\n",
    "print(\"Name: \" + record.name)\n",
    "print(\"Lenght:\" + str(len(record.seq)))\n",
    "print(\"Type: \" + record.annotations[\"molecule_type\"] + \" \" +  record.annotations[\"topology\"])\n",
    "print(\"Accessions\" + str(record.annotations[\"accessions\"]))\n",
    "print(\"Source: \" + record.annotations[\"source\"])\n",
    "print()\n",
    "print(\"Publications:\")\n",
    "for c in record.annotations[\"references\"]:\n",
    "    print(c.title)\n",
    "    print(c.journal)\n",
    "print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HIVI5C'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID = record.name\n",
    "Organism = record.annotations[\"organism\"]\n",
    "Accession = record.annotations[\"accessions\"][0]\n",
    "Definition = record.description\n",
    "Keywords = \"\"\n",
    "\n",
    "for c in record.annotations[\"keywords\"]:\n",
    "    Keywords += c + \", \"\n",
    "Keywords = Keywords[0: (len(Keywords) -2)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HIVI5C\n",
      "Human immunodeficiency virus 1\n",
      "L42022\n",
      "Human immunodeficiency virus type 1 (isolate genotype C, I5) gag gene, partial cds\n",
      "gag gene, p24 protein\n"
     ]
    }
   ],
   "source": [
    "print(ID)\n",
    "print(Organism)\n",
    "print(Accession)\n",
    "print(Definition)\n",
    "print(Keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "231\n",
      "DNA\n",
      "VRL\n",
      "24-MAR-1997\n"
     ]
    }
   ],
   "source": [
    "Size = len(record.seq)\n",
    "print(Size)\n",
    "Molecular_type = record.annotations[\"molecule_type\"]\n",
    "print(Molecular_type)\n",
    "Genbank_division = record.annotations[\"data_file_division\"]\n",
    "print(Genbank_division)\n",
    "Modification_date = record.annotations[\"date\"]\n",
    "print(Modification_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATCCAGTACATGCAGGGCCTATTGCACCAGGCCAAATGAGAGAACCAAGGGGAAGTGACATAGCAGGAACTACAAGTACCCTTCAGGAACAAGTAGCATGGATGACAGGTAACCCACCTGTTCCAGTGGGAGAAATCTATAAAAGATGGATAATTCTGGGATTAAATAAAATAGTAAGAATGTATAGCCCTGTCAGCATTTTGGACATAAAACAAGGGCCAAAGGAACCC\n",
      "86\n",
      "46\n",
      "44\n",
      "55\n"
     ]
    }
   ],
   "source": [
    "Sequence = record.seq\n",
    "print(Sequence)\n",
    "Count_A = record.seq.count(\"A\")\n",
    "print(Count_A)\n",
    "Count_C = record.seq.count(\"C\")\n",
    "print(Count_C)\n",
    "Count_T = record.seq.count(\"T\")\n",
    "print(Count_T)\n",
    "Count_G = record.seq.count(\"G\")\n",
    "print(Count_G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "location: [0:231]\n",
      "authors: Voevodin,A., Crandall,K.A., Seth,P. and al Mufti,S.\n",
      "title: HIV type 1 subtypes B and C from new regions of India and Indian and Ethiopian expatriates in Kuwait\n",
      "journal: AIDS Res. Hum. Retroviruses 12 (7), 641-643 (1996)\n",
      "medline id: \n",
      "pubmed id: 8743090\n",
      "comment: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# CDS = record.features\n",
    "# print(CDS)   \n",
    "Pubmed = \"\"  \n",
    "for c in record.annotations[\"references\"]:\n",
    "    Pubmed += str(c)\n",
    "    \n",
    "print(Pubmed)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (513129283.py, line 25)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  Input \u001b[1;32mIn [63]\u001b[1;36m\u001b[0m\n\u001b[1;33m    cur.execute(\"\"\"\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mIndentationError\u001b[0m\u001b[1;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Locus (ID, Size, Molecular_type, Genbank_division, Modification_date)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Size, Molecular_type, Genbank_division, Modification_date))\n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Sequences (ID, Sequence, Count_A, Count_C, Count_T, Count_G)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s,%s)\n",
    "        \"\"\",\n",
    "        (ID, Sequence, Count_A, Count_C, Count_T, Count_G))   #retirei o length porque j√° tem no locus\n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Pubmed_Info (ID_pubmed, Title, Abstract, Pubmed_Infocol, Authors)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID_pubmed, Title, Abstract, Pubmed_Infocol, Authors))   \n",
    "    \n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO References (ID, Pubmed, Title, Journal, Consortium, Remark)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID, Pubmed, Title, Journal, Consortium, Remark))   \n",
    "\n",
    "     cur.execute(\"\"\"\n",
    "        INSERT INTO Features (ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA)    \n",
    "        VALUES (%s, %s, %s, %s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID, CDS, Gene, Regulatory, exons, poli_A-site, misc_feature, mRNA))   \n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO CDS (ID_CDS, Translacion_seq, Localization, ID_protein, Protein)    \n",
    "        VALUES (%s, %s, %s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID_CDS, Translacion_seq, Localization, ID_protein, Protein)) \n",
    "\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Source (ID, Location_span)    \n",
    "        VALUES (%s, %s))\n",
    "        \"\"\",\n",
    "        (ID, Location_span)) \n",
    "    \n",
    "     cur.execute(\"\"\"\n",
    "        INSERT INTO Authors (ID_authors, Name, Affiliation)    \n",
    "        VALUES (%s, %s, %s))\n",
    "        \"\"\",\n",
    "        (ID_authors, Name, Affiliation))    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "\n",
    "\n",
    "conn = psycopg2.connect('postgresql://geobox:xxxxxxxxxxxx@geo.di.uminho.pt/bioinformatica')\n",
    "conn.autocommit = True\n",
    "\n",
    "cur = conn.cursor()\n",
    "try:\n",
    "    cur.execute(\"\"\"\n",
    "        INSERT INTO Genbank (ID, Organism, Accession, Definition, Keywords)    \n",
    "        VALUES (%s, %s, %s, %s, %s)\n",
    "        \"\"\",\n",
    "        (ID, Organism, Accession, Definition, Keywords))\n",
    "        \n",
    "        \n",
    "except psycopg2.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "finally:\n",
    "    conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "12752d1325306e0d13ba81c4f4222c0600fa831e8bd8a6fa39f7f9f637486b90"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
