{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Seq:\n",
    "    \"\"\" \n",
    "    Classe usada para manipular e transformar os diferentes tipos de sequências através de diferentes processos.\n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self, seq):\n",
    "        \"\"\"Construtor da classe Seq.\n",
    "\n",
    "        Args:\n",
    "            seq (str): Sequência biológica.\n",
    "        \"\"\"\n",
    "        self.seq = str(seq.upper())\n",
    "\n",
    "    def __len__ (self): \n",
    "        \"\"\" Função que permite implementar a função len().\n",
    "\n",
    "        Returns:\n",
    "            int: Tamanho da sequência.\n",
    "        \"\"\"   \n",
    "        return len(self.seq)\n",
    "    \n",
    "    def __str__(self):\n",
    "        \"\"\" Função que permite representar a sequência como uma string.\n",
    "\n",
    "        Returns:\n",
    "            str: Representação da sequência.      \n",
    "        \"\"\"       \n",
    "        return self.seq\n",
    "\n",
    "    def __getitem__(self,n):\n",
    "        \"\"\"Função que permite aplicar a indexação na classe.\n",
    "\n",
    "        Args:\n",
    "            n (int): Index\n",
    "\n",
    "        Returns:\n",
    "            str: Base ou aminoácido correspondente ao index.\n",
    "        \"\"\"\n",
    "        return self.seq[n]\n",
    "    \n",
    "    def __getslice__(self,i,j):\n",
    "        \"\"\"Função que permite aplicar a indexação na classe.\n",
    "\n",
    "        Args:\n",
    "            i (int): index start\n",
    "            j (int): index final\n",
    "\n",
    "        Returns:\n",
    "            str: Sequência, base ou aminoácido correspondente ao index.\n",
    "        \"\"\"\n",
    "        return self.seq[a]    \n",
    "\n",
    "    def get_type (self):\n",
    "        \"\"\"Função que a partir de uma sequência, classifica-a em DNA, RNA ou AMINO ACID SEQUENCE.\n",
    "\n",
    "        Returns:\n",
    "            str: Classificação da sequência.\n",
    "        \"\"\"\n",
    "        if len([c for c in self.seq if c not in \"ACGT\"]) == 0:\n",
    "            return \"DNA\"\n",
    "\n",
    "        elif len([c for c in self.seq if c not in \"ACGU\"]) == 0:\n",
    "            return \"RNA\"\n",
    "\n",
    "        elif len([c for c in self.seq if c not in \"ABCDEFGHIKLMNPQRSTVWYZ_\"]) == 0:\n",
    "            return \"AMINO ACID SEQUENCE\"\n",
    "        \n",
    "        else:\n",
    "            return \"Not valid\"\n",
    "\n",
    "    def count_nucleotides (self):\n",
    "        \"\"\"Função que a partir de uma sequência de DNA ou RNA e conta o número dos diferentes nucleótidos presentes.\n",
    "\n",
    "        Returns: \n",
    "            list: Lista com a contagem dos diferentes núcleotidos.           \n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"DNA\":\n",
    "            A = self.seq.count(\"A\")\n",
    "            C = self.seq.count(\"C\")\n",
    "            G = self.seq.count(\"G\")\n",
    "            T = self.seq.count(\"T\") \n",
    "            return [A, C, G, T]\n",
    "        \n",
    "        elif Seq(self.seq).get_type() == \"RNA\":                    \n",
    "            A = self.seq.count(\"A\")\n",
    "            C = self.seq.count(\"C\")\n",
    "            G = self.seq.count(\"G\")\n",
    "            U = self.seq.count(\"U\") \n",
    "            return [A, C, G, U] \n",
    "        else:\n",
    "            return \"Not DNA or RNA\"\n",
    "\n",
    "    def freq_nucleotides (self):\n",
    "        \"\"\"Função que a partir de uma sequência de DNA ou RNA, estima a frequência dos diferentes nucleótidos presentes.\n",
    "\n",
    "        Returns:\n",
    "            list: Lista com a frequência dos diferentes nucleótidos. \n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"DNA\":\n",
    "            A = self.seq.count(\"A\") / len(self.seq)\n",
    "            C = self.seq.count(\"C\") / len(self.seq)\n",
    "            G = self.seq.count(\"G\") / len(self.seq)\n",
    "            T = self.seq.count(\"T\") / len(self.seq)\n",
    "            return [round(A, 3), round(C, 3), round(G, 3), round(T, 3)]             \n",
    "        \n",
    "        elif Seq(self.seq).get_type() == \"RNA\":\n",
    "            A = self.seq.count(\"A\") / len(self.seq)\n",
    "            C = self.seq.count(\"C\") / len(self.seq)\n",
    "            G = self.seq.count(\"G\") / len(self.seq)\n",
    "            U = self.seq.count(\"U\") / len(self.seq)\n",
    "            return [round(A, 3), round(C, 3), round(G, 3), round(U, 3)]\n",
    "        else:\n",
    "            return \"Not DNA or RNA\"\n",
    "\n",
    "    def transcricao(self):\n",
    "        \"\"\"Função que a partir de uma sequência de DNA ou RNA, devolve o transcrito da mesma.\n",
    "\n",
    "        Returns:\n",
    "            str: Sequência do transcrito.\n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"DNA\":  \n",
    "           DNA = self.seq.lower() \n",
    "           seq_trans = DNA.replace(\"a\",\"U\").replace(\"u\",\"A\").replace(\"c\",\"G\").replace(\"g\",\"C\")[::-1].upper()\n",
    "           return seq_trans  \n",
    "\n",
    "        elif Seq(self.seq).get_type() == \"RNA\": \n",
    "           RNA = self.seq.lower() \n",
    "           seq_trans = RNA.replace(\"u\",\"A\").replace(\"a\",\"T\").replace(\"g\",\"C\").replace(\"c\",\"G\")[::-1].upper()\n",
    "           return seq_trans  \n",
    "\n",
    "        else:\n",
    "            return \"Not DNA or RNA\" \n",
    "\n",
    "    def get_codons (self):\n",
    "        \"\"\"Função que recebe uma sequência de DNA ou RNA e transforma cada conjunto de 3 nucleótidos, presentes ao longo da sequência, no respetivo codão.  \n",
    "\n",
    "        Returns:\n",
    "            list: Lista de codões.\n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"DNA\" or Seq(self.seq).get_type() == \"RNA\":\n",
    "            l = []\n",
    "            codon_list =[]\n",
    "            for c in range (0, len(self.seq), 3):                             \n",
    "                codon_list.append(self.seq[c : c + 3])\n",
    "            if len(codon_list[-1]) < 3:\n",
    "                codon_list.pop()\n",
    "            return codon_list\n",
    "        else:\n",
    "            return \"Not DNA or RNA\"   \n",
    "\n",
    "    def traducao(self):\n",
    "        \"\"\"\"Função que a partir de uma sequência de DNA, origina uma sequência de aminoácidos. \n",
    "\n",
    "        Returns:\n",
    "            str: Sequência de aminoácidos\n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"DNA\":                                 \n",
    "            gencode = {\n",
    "            'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "            'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "            'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "            'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "            'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "            'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "            'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "            'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "            'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "            'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "            'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "            'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "            'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "            'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "            'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "            'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n",
    "        \n",
    "            l = []\n",
    "            for c in Seq(self.seq).get_codons():\n",
    "                l.append(gencode.get(c))\n",
    "                l_final = \"\".join(l)    \n",
    "            return l_final  \n",
    "        else:\n",
    "            return \"Not DNA\"\n",
    "\n",
    "    def get_prots (self):   \n",
    "        \"\"\"Função que a partir de uma sequência de aminoácidos gera as diferentes proteínas.\n",
    "\n",
    "        Returns:\n",
    "            list: Lista de proteínas\n",
    "        \"\"\"\n",
    "        if Seq(self.seq).get_type() == \"AMINO ACID SEQUENCE\":\n",
    "            lista_final = []\n",
    "            lista_stop_tratados = self.seq.split(\"_\") \n",
    "            for c in lista_stop_tratados:        \n",
    "                for idx, amino_s in enumerate(c):       \n",
    "                    if amino_s == \"M\":                  \n",
    "                        lista_final.append(c[idx:]) \n",
    "            return lista_final \n",
    "        else:\n",
    "            return \"not AMINO ACID SEQUENCE\"\n",
    "\n",
    "    def get_orfs (self):    \n",
    "        \"\"\"Função que a partir de uma sequência de DNA ou RNA origina diferentes tipos de ORFS. \n",
    "\n",
    "        Returns:\n",
    "            list: Lista das ORFS\n",
    "        \"\"\"\n",
    "        if  Seq(self.seq).get_type() == \"DNA\": \n",
    "            dna = self.seq.lower()\n",
    "            dna_f = self.seq\n",
    "            lista_orfs =[]\n",
    "            lista_orfs.append(dna_f[0:])   \n",
    "            lista_orfs.append(dna_f[1:])   \n",
    "            lista_orfs.append(dna_f[2:])\n",
    "            dna_rev_comp = dna.replace(\"a\",\"T\").replace(\"t\",\"A\").replace(\"c\",\"G\").replace(\"g\",\"C\")[::-1] \n",
    "            lista_orfs.append(dna_rev_comp[0:])\n",
    "            lista_orfs.append(dna_rev_comp[1:])  \n",
    "            lista_orfs.append(dna_rev_comp[2:])\n",
    "            return lista_orfs\n",
    "        elif Seq(self.seq).get_type() == \"RNA\":\n",
    "            rna = self.seq.lower()\n",
    "            rna_f = self.seq\n",
    "            lista_orfs =[]\n",
    "            lista_orfs.append(rna_f[0:])      \n",
    "            lista_orfs.append(rna_f[1:])   \n",
    "            lista_orfs.append(rna_f[2:])\n",
    "            rna_rev_comp = rna.replace(\"a\",\"U\").replace(\"u\",\"A\").replace(\"c\",\"G\").replace(\"g\",\"C\")[::-1] \n",
    "            lista_orfs.append(rna_rev_comp[0:])\n",
    "            lista_orfs.append(rna_rev_comp[1:]) \n",
    "            lista_orfs.append(rna_rev_comp[2:])\n",
    "            return lista_orfs\n",
    "        else:\n",
    "            return \"Not DNA or RNA\"\n",
    "\n",
    "    def get_all_prots(self):\n",
    "        \"\"\"Função que a partir de uma sequência de DNA e permite obter todas as possíveis proteínas. \n",
    "\n",
    "        Returns:\n",
    "            list: Lista das proteínas\n",
    "        \"\"\"\n",
    "        if  Seq(self.seq).get_type() == \"DNA\": \n",
    "            orfs = Seq.get_orfs(self)\n",
    "            list_codons = []\n",
    "            list_aminos = []\n",
    "            list_prots =  []\n",
    "            \n",
    "            for c  in orfs:\n",
    "                a = Seq(c).get_codons()\n",
    "                list_codons.append(a)\n",
    "                for e in list_codons:\n",
    "                    b = Seq(c).traducao()\n",
    "                list_aminos.append((b))\n",
    "                for r in list_aminos:\n",
    "                    c = Seq(r).get_prots()\n",
    "                list_prots.append(c)       \n",
    "            return [c for a in list_prots for c in a]\n",
    "        else:\n",
    "            return \"Not DNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "class Test_Seq(unittest.TestCase):\n",
    "    \n",
    "    def test_get_type(self):\n",
    "        \"\"\"Função usada para testar a função get_type() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATGC\").get_type(),\"DNA\")\n",
    "        self.assertEqual(Seq(\"augc\").get_type(),\"RNA\")\n",
    "        self.assertEqual(Seq(\"MNAAQTHCTA\").get_type(),\"AMINO ACID SEQUENCE\")\n",
    "        self.assertEqual(Seq(\"A-G-T-A-T\").get_type(),\"Not valid\")\n",
    "        self.assertEqual(Seq(\"ATGC1\").get_type(),\"Not valid\")\n",
    "\n",
    "    def test_count_nucleotides(self):\n",
    "        \"\"\"Função usada para testar a função count_nucleotides() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").count_nucleotides(),[1, 1, 1, 3])\n",
    "        self.assertEqual(Seq(\"auuugc\").count_nucleotides(),[1, 1, 1, 3]) \n",
    "        self.assertEqual(Seq(\"AUTUUGC\").count_nucleotides(),\"Not DNA or RNA\") \n",
    "\n",
    "    def test_freq_nucleotides(self):\n",
    "        \"\"\"Função usada para testar a função freq_nucleotides() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").freq_nucleotides(),[0.167, 0.167, 0.167, 0.5])\n",
    "        self.assertEqual(Seq(\"auuagc\").freq_nucleotides(),[0.333, 0.167, 0.167, 0.333])\n",
    "        self.assertEqual(Seq(\"AT TAC-AC\").freq_nucleotides(),\"Not DNA or RNA\")\n",
    "        \n",
    "    def test_transcricao(self):\n",
    "        \"\"\"Função usada para testar a função transcricao() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").transcricao(),\"GCTTTU\")\n",
    "        self.assertEqual(Seq(\"auuagc\").transcricao(),\"GCTAAT\")\n",
    "        self.assertEqual(Seq(\"AUUHAGC\").transcricao(),\"Not DNA or RNA\")\n",
    "    \n",
    "    def test_get_codons(self):\n",
    "        \"\"\"Função usada para testar a função get_codons() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_codons(),['ATT', 'TGC'])\n",
    "        self.assertEqual(Seq(\"auuagc\").get_codons(),['AUU', 'AGC'])\n",
    "        self.assertEqual(Seq(\"acacc_ca\").get_codons(),\"Not DNA or RNA\")\n",
    "        self.assertEqual(Seq(\"ATTTGC1\").get_codons(),\"Not DNA or RNA\")\n",
    "    \n",
    "    def test_traducao(self):\n",
    "        \"\"\"Função usada para testar a função traducao() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").traducao(),\"IC\")\n",
    "        self.assertEqual(Seq(\"auuagc\").traducao(),\"Not DNA\")\n",
    "\n",
    "    def test_get_prots(self):\n",
    "        \"\"\"Função usada para testar a função get_prots() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"MA_KLCMNB\").get_prots(),['MA', 'MNB'])\n",
    "        self.assertEqual(Seq(\"ATTMTGC\").get_prots(),['MTGC'])\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_prots(),\"not AMINO ACID SEQUENCE\")\n",
    "        self.assertEqual(Seq(\"auuagc\").get_prots(),\"not AMINO ACID SEQUENCE\")\n",
    "        \n",
    "    def test_get_orfs(self):\n",
    "        \"\"\"Função usada para testar a função get_orfs() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_orfs(),['ATTTGC', 'TTTGC', 'TTGC', 'GCAAAT', 'CAAAT', 'AAAT'])\n",
    "        self.assertEqual(Seq(\"auuagc\").get_orfs(),['AUUAGC', 'UUAGC', 'UAGC', 'GCUAAU', 'CUAAU', 'UAAU'])\n",
    "        self.assertEqual(Seq(\"MA_KLCMNB\").get_orfs(),\"Not DNA or RNA\")\n",
    "        self.assertEqual(Seq(\"ATTTGC1\").get_orfs(),\"Not DNA or RNA\")\n",
    "\n",
    "    def test_get_all_prots(self):\n",
    "        \"\"\"Função usada para testar a função get_all_prots() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_all_prots(),[])\n",
    "        self.assertEqual(Seq(\"ATTATGTGC\").get_all_prots(),['MC'])\n",
    "        self.assertEqual(Seq(\"auguagc\").get_all_prots(),\"Not DNA\")\n",
    "    \n",
    "unittest.main(argv=[\"\"], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Mat:\n",
    "    \"\"\"\n",
    "    Classe que permite criar e manipular Matrizes\n",
    "   \n",
    "    \"\"\"\n",
    "    def __init__(self, rows, cols):\n",
    "        \"\"\"Construtor da Matriz \n",
    "\n",
    "        Args:\n",
    "            rows (int): Número de linhas\n",
    "            cols (int): Número de colunas \n",
    "        \"\"\"\n",
    "        self.mat = [[0 for c in range(cols)]\n",
    "                    for r in range(rows)]\n",
    "    \n",
    "\n",
    "    def numRows (self): \n",
    "        \"\"\"Função que permite obter o número de linhas da Matriz\n",
    "\n",
    "        Returns:\n",
    "            int : Valor de linhas da Matriz\n",
    "        \"\"\"\n",
    "        return len(self.mat)\n",
    "\n",
    "\n",
    "    def numCols (self): \n",
    "        \"\"\"Função que permite obter o númeor de colunas da Matriz\n",
    "\n",
    "        Returns:\n",
    "            int : Valor de colunas da Matriz\n",
    "        \"\"\"\n",
    "        return len(self.mat[0])\n",
    "\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\" Função que permite devolver a Matriz como uma string\n",
    "\n",
    "        Returns:\n",
    "            str : String da Matriz\n",
    "        \"\"\"\n",
    "        return '\\n'.join(' '.join(str(val) for val in row)\n",
    "                         for row in self.mat)\n",
    "\n",
    "    \n",
    "    def __getitem__ (self, n):\n",
    "        return self.mat[n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "def subst(x, y):\n",
    "  import blosum as bl\n",
    "\n",
    "  \"\"\"Funçao que permite obter o valor da substituição do Aminoácido da Blossum 62\n",
    "\n",
    "  Args:\n",
    "      x (str): Aminoácido 1\n",
    "      y (str): Aminoácido 2\n",
    "\n",
    "  Returns:\n",
    "      int: Valor de substituição do Aminoácido \n",
    "  \"\"\"\n",
    "  \n",
    "  dic = dict(bl.BLOSUM(62))\n",
    "  return int(dic[x+y])   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NW:\n",
    "  \"\"\"Classe que permite efetuar o Algoritmo de Needleman Wunsch de forma a realizar o alinhamento Global de um par de sequências.\n",
    "  \"\"\"\n",
    "\n",
    "  def __init__(self, s1, s2, g = -4):\n",
    "    \"\"\"Contrutor da classe NW\n",
    "\n",
    "    Args:\n",
    "        s1 (str): Sequência 1\n",
    "        s2 (str): Sequência 2\n",
    "        g (int, optional): Gap Penalty. Defaults to -4.\n",
    "    \"\"\"\n",
    "    self.s1 = s1\n",
    "    self.s2 = s2\n",
    "    self.mat = Mat(len(s1) + 1, len(s2) + 1)  \n",
    "    self.tr  = Mat(len(s1) + 1, len(s2) + 1)\n",
    "    \n",
    "    for L in range(len(s1)):                     \n",
    "      self.mat[L + 1][0] = g * (L + 1)           \n",
    "      self.tr[L + 1][0]  = 'C'                      \n",
    "\n",
    "    for C in range(len(s2)):                     \n",
    "      self.mat[0][C + 1] = g * (C + 1)             \n",
    "      self.tr[0][C + 1]  = 'E'                        \n",
    "\n",
    "    for L, x1 in enumerate(s1):\n",
    "      for C, x2 in enumerate(s2):\n",
    "        possiveis = [\n",
    "            self.mat[L  ][C    ] + subst(x1, x2),   \n",
    "            self.mat[L+1][C    ] + g,               \n",
    "            self.mat[L  ][C + 1] + g,               \n",
    "        ]\n",
    "        dirs = \"DEC\"\n",
    "\n",
    "        self.mat[L + 1][C + 1] = max(possiveis)   \n",
    "        self.tr[L + 1][C + 1] = dirs[possiveis.index(self.mat[L + 1][C + 1])]  \n",
    "\n",
    "  def rebuild(self):\n",
    "    \"\"\"Função que permite obter um possivel alinhamento das sequências\n",
    "\n",
    "    Returns:\n",
    "        str: Sequências alinhadas \n",
    "    \"\"\"\n",
    "    L = len(self.s1)\n",
    "    C = len(self.s2)\n",
    "    S1 = \"\"\n",
    "    S2 = \"\"\n",
    "    \n",
    "    dirs = {\n",
    "        'D' : (-1, -1),\n",
    "        'E' : ( 0, -1),\n",
    "        'C' : (-1,  0)\n",
    "    }\n",
    "\n",
    "    while L > 0 or C > 0:\n",
    "\n",
    "      DL, DC = dirs[self.tr[L][C]]\n",
    "\n",
    "      if self.tr[L][C] == \"D\":\n",
    "        S1 = self.s1[L - 1] + S1    \n",
    "        S2 = self.s2[C - 1] + S2\n",
    "      elif self.tr[L][C] == \"E\":\n",
    "        S1 = '-' + S1\n",
    "        S2 = self.s2[C - 1] + S2\n",
    "      else:\n",
    "        S1 = self.s1[L - 1] + S1\n",
    "        S2 = '-' + S2        \n",
    "\n",
    "      L += DL\n",
    "      C += DC\n",
    "\n",
    "    return S1, S2\n",
    "\n",
    " \n",
    "  def max_score (self):\n",
    "    \"\"\"Função que permite obter o max score do alinhamento \n",
    "\n",
    "    Returns:\n",
    "        int: Max Score\n",
    "    \"\"\"\n",
    "    return self.mat[len(self.s1)][len(self.s2)]\n",
    "\n",
    "  def __repr__(self):\n",
    "    \"\"\"Função que permite obter a representação em String da própria classe \n",
    "\n",
    "    Returns:\n",
    "        str: Matriz S (Score) e Matriz T (Trace)\n",
    "    \"\"\"\n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, sep = '   ', file = S)           \n",
    "      for L, linha in zip(lins, self.mat):\n",
    "        print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "\n",
    "      print(file = S)\n",
    "\n",
    "      print(' ', *cols, file = S)\n",
    "      for L, linha in zip(lins, self.tr):\n",
    "        print(L, *linha, file = S)\n",
    "\n",
    "      return S.getvalue()\n",
    "\n",
    "  def score_mat(self):\n",
    "    \"\"\"Função que permite obter a Matriz S (Score)\n",
    "\n",
    "    Returns:\n",
    "        str: Matriz S\n",
    "    \"\"\"\n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, sep = '   ', file = S)           \n",
    "      for L, linha in zip(lins, self.mat):\n",
    "        print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "\n",
    "      print(file = S)\n",
    "      return print(S.getvalue())\n",
    "\n",
    "  def trace_mat(self):\n",
    "    \"\"\"Função que permite obter a Matriz T (Trace)\n",
    "\n",
    "    Returns:\n",
    "        str: Matriz T (Trace)\n",
    "    \"\"\"\n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, file = S)\n",
    "      for L, linha in zip(lins, self.tr):\n",
    "        print(L, *linha, file = S)\n",
    "\n",
    "      return print(S.getvalue())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "class Test_Seq(unittest.TestCase):\n",
    "    \n",
    "    def test_get_type(self):\n",
    "        \"\"\"Função usada para testar a função get_type() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATGC\").get_type(),\"DNA\")\n",
    "        self.assertEqual(Seq(\"augc\").get_type(),\"RNA\")\n",
    "        self.assertEqual(Seq(\"MNAAQTHCTA\").get_type(),\"AMINO ACID SEQUENCE\")\n",
    "        self.assertEqual(Seq(\"A-G-T-A-T\").get_type(),\"Not valid\")\n",
    "        self.assertEqual(Seq(\"ATGC1\").get_type(),\"Not valid\")\n",
    "\n",
    "    def test_count_nucleotides(self):\n",
    "        \"\"\"Função usada para testar a função count_nucleotides() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").count_nucleotides(),[1, 1, 1, 3])\n",
    "        self.assertEqual(Seq(\"auuugc\").count_nucleotides(),[1, 1, 1, 3]) \n",
    "        self.assertEqual(Seq(\"AUTUUGC\").count_nucleotides(),\"Not DNA or RNA\") \n",
    "\n",
    "    def test_freq_nucleotides(self):\n",
    "        \"\"\"Função usada para testar a função freq_nucleotides() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").freq_nucleotides(),[0.167, 0.167, 0.167, 0.5])\n",
    "        self.assertEqual(Seq(\"auuagc\").freq_nucleotides(),[0.333, 0.167, 0.167, 0.333])\n",
    "        self.assertEqual(Seq(\"AT TAC-AC\").freq_nucleotides(),\"Not DNA or RNA\")\n",
    "        \n",
    "    def test_transcricao(self):\n",
    "        \"\"\"Função usada para testar a função transcricao() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").transcricao(),\"GCTTTU\")\n",
    "        self.assertEqual(Seq(\"auuagc\").transcricao(),\"GCTAAT\")\n",
    "        self.assertEqual(Seq(\"AUUHAGC\").transcricao(),\"Not DNA or RNA\")\n",
    "    \n",
    "    def test_get_codons(self):\n",
    "        \"\"\"Função usada para testar a função get_codons() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_codons(),['ATT', 'TGC'])\n",
    "        self.assertEqual(Seq(\"auuagc\").get_codons(),['AUU', 'AGC'])\n",
    "        self.assertEqual(Seq(\"acacc_ca\").get_codons(),\"Not DNA or RNA\")\n",
    "        self.assertEqual(Seq(\"ATTTGC1\").get_codons(),\"Not DNA or RNA\")\n",
    "    \n",
    "    def test_traducao(self):\n",
    "        \"\"\"Função usada para testar a função traducao() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").traducao(),\"IC\")\n",
    "        self.assertEqual(Seq(\"auuagc\").traducao(),\"Not DNA\")\n",
    "\n",
    "    def test_get_prots(self):\n",
    "        \"\"\"Função usada para testar a função get_prots() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"MA_KLCMNB\").get_prots(),['MA', 'MNB'])\n",
    "        self.assertEqual(Seq(\"ATTMTGC\").get_prots(),['MTGC'])\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_prots(),\"not AMINO ACID SEQUENCE\")\n",
    "        self.assertEqual(Seq(\"auuagc\").get_prots(),\"not AMINO ACID SEQUENCE\")\n",
    "        \n",
    "    def test_get_orfs(self):\n",
    "        \"\"\"Função usada para testar a função get_orfs() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_orfs(),['ATTTGC', 'TTTGC', 'TTGC', 'GCAAAT', 'CAAAT', 'AAAT'])\n",
    "        self.assertEqual(Seq(\"auuagc\").get_orfs(),['AUUAGC', 'UUAGC', 'UAGC', 'GCUAAU', 'CUAAU', 'UAAU'])\n",
    "        self.assertEqual(Seq(\"MA_KLCMNB\").get_orfs(),\"Not DNA or RNA\")\n",
    "        self.assertEqual(Seq(\"ATTTGC1\").get_orfs(),\"Not DNA or RNA\")\n",
    "\n",
    "    def test_get_all_prots(self):\n",
    "        \"\"\"Função usada para testar a função get_all_prots() da classe Seq.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Seq(\"ATTTGC\").get_all_prots(),[])\n",
    "        self.assertEqual(Seq(\"ATTATGTGC\").get_all_prots(),['MC'])\n",
    "        self.assertEqual(Seq(\"auguagc\").get_all_prots(),\"Not DNA\")\n",
    "    \n",
    "unittest.main(argv=[\"\"], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "..\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 0.004s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b604c49d0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "\n",
    "class Test_NW(unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a Classe NW\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    def test_rebuild (self):\n",
    "        \"\"\"Testar a função rebuild\n",
    "        \"\"\"\n",
    "\n",
    "        self.assertEqual(NW(\"ACTG\", \"ACTG\").rebuild(), ('ACTG', 'ACTG'))\n",
    "        self.assertEqual(NW(\"ATCAT\", \"ACAT\").rebuild(), ('ATCAT', 'A-CAT'))\n",
    "\n",
    "\n",
    "    def test_max_score (self):\n",
    "        \"\"\"Testar a função max_score\n",
    "        \"\"\"\n",
    "\n",
    "        self.assertEqual(NW(\"ACTG\", \"ACTG\").max_score(), 24)\n",
    "        self.assertEqual(NW(\"ATCAT\", \"ACAT\").max_score(), 18)\n",
    "\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SW:\n",
    "  \"\"\"Classe que permite efetuar o Algoritmo de Smith Waterman de forma a realizar o alinhamento Local de um par de sequências.\n",
    "  \"\"\"\n",
    "\n",
    "\n",
    "  def __init__(self, s1, s2, g = -4):\n",
    "    \"\"\"Contrutor da classe SW\n",
    "\n",
    "    Args:\n",
    "        s1 (str): Sequência 1\n",
    "        s2 (str): Sequência 2\n",
    "        g (int, optional): Gap Penalty. Defaults to -4.\n",
    "    \"\"\"\n",
    "\n",
    "    self.s1 = s1\n",
    "    self.s2 = s2\n",
    "    self.mat = Mat(len(s1) + 1, len(s2) + 1) \n",
    "    self.tr  = Mat(len(s1) + 1, len(s2) + 1)\n",
    "\n",
    "\n",
    "    for L, x1 in enumerate(s1):\n",
    "      for C, x2 in enumerate(s2):\n",
    "        possiveis = [\n",
    "            self.mat[L  ][C    ] + subst(x1, x2),   \n",
    "            self.mat[L+1][C    ] + g,               \n",
    "            self.mat[L  ][C + 1] + g,               \n",
    "            0]\n",
    "        dirs = \"DEC.\"\n",
    "\n",
    "        self.mat[L + 1][C + 1] = max(possiveis)   \n",
    "        if self.mat[L + 1][C + 1] != 0:\n",
    "          self.tr[L + 1][C + 1] = dirs[possiveis.index(self.mat[L + 1][C + 1])] \n",
    "  \n",
    "  def max_score (self):\n",
    "    \"\"\"Função que permite obter o max score do alinhamento \n",
    "\n",
    "    Returns:\n",
    "        int: Valor do Max Score\n",
    "    \"\"\"\n",
    "    max_score = 0\n",
    "    for L, x1 in enumerate(self.mat):\n",
    "      for C, x2 in enumerate(self.mat):\n",
    "        if max(x1) > max_score:\n",
    "          max_score = max(x1)\n",
    "    return max_score\n",
    "\n",
    "  \n",
    "  def rebuild(self):\n",
    "    \"\"\"Função que permite obter um possivel alinhamento das sequências\n",
    "\n",
    "    Returns:\n",
    "        str: Sequências alinhadas \n",
    "    \"\"\"\n",
    "    max_score = 0\n",
    "    linha = 0\n",
    "    coluna = 0\n",
    "    for L, x1 in enumerate(self.mat):\n",
    "      for C, x2 in enumerate(self.mat):\n",
    "        if max(x1) > max_score:\n",
    "          max_score = max(x1)\n",
    "          linha = L\n",
    "        coluna = C\n",
    "    coluna -= 1\n",
    "    L = linha\n",
    "    C = coluna + 1\n",
    "    S1 = \"\"\n",
    "    S2 = \"\"\n",
    "    \n",
    "    dirs = {\n",
    "        'D' : (-1, -1),\n",
    "        'E' : ( 0, -1),\n",
    "        'C' : (-1,  0),\n",
    "        \".\" : (0, 0)\n",
    "    }\n",
    "\n",
    "    while L >= 0 or C >= 0:\n",
    "\n",
    "      try:\n",
    "        DL, DC = dirs[self.tr[L][C]]\n",
    "        if self.tr[L][C] == \"D\":\n",
    "            S1 = self.s1[L - 1] + S1    \n",
    "            S2 = self.s2[C - 1] + S2\n",
    "        elif self.tr[L][C] == \"E\":\n",
    "            S1 = '-' + S1\n",
    "            S2 = self.s2[C - 1] + S2\n",
    "        elif self.tr[L][C] == \"C\":\n",
    "            S1 = self.s1[L - 1] + S1\n",
    "            S2 = '-' + S2 \n",
    "        else:\n",
    "          break\n",
    "    \n",
    "        L += DL\n",
    "        C += DC\n",
    "      except:\n",
    "        break\n",
    "\n",
    "    \n",
    "    return S1, S2\n",
    "\n",
    " \n",
    "\n",
    "        \n",
    "\n",
    "    \n",
    "\n",
    "  def __repr__(self):\n",
    "    \"\"\"Função que permite obter a representação em String da própria classe \n",
    "\n",
    "    Returns:\n",
    "        str: Matriz S (Score) e Matriz T (Trace)\n",
    "    \"\"\"\n",
    "    \n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, sep = '   ', file = S)           \n",
    "      for L, linha in zip(lins, self.mat):\n",
    "        print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "\n",
    "      print(file = S)\n",
    "\n",
    "      print(' ', *cols, file = S)\n",
    "      for L, linha in zip(lins, self.tr):\n",
    "        print(L, *linha, file = S)\n",
    "\n",
    "      return S.getvalue()\n",
    "\n",
    "  def score_mat(self):\n",
    "    \"\"\"Função que permite obter a Matriz S (Score)\n",
    "\n",
    "    Returns:\n",
    "        str: Matriz S\n",
    "    \"\"\"\n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, sep = '   ', file = S)           \n",
    "      for L, linha in zip(lins, self.mat):\n",
    "        print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "\n",
    "      print(file = S)\n",
    "      return print(S.getvalue())\n",
    "\n",
    "  def trace_mat(self):\n",
    "    \"\"\"Função que permite obter a Matriz T (Trace)\n",
    "\n",
    "    Returns:\n",
    "        str: Matriz T (Trace)\n",
    "    \"\"\"\n",
    "    cols = \"-\" + self.s2\n",
    "    lins = \"-\" + self.s1\n",
    "    with io.StringIO(\"\") as S:\n",
    "      print(' ', *cols, file = S)\n",
    "      for L, linha in zip(lins, self.tr):\n",
    "        print(L, *linha, file = S)\n",
    "\n",
    "      return print(S.getvalue())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....\n",
      "----------------------------------------------------------------------\n",
      "Ran 4 tests in 0.010s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b604c4ac0>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "\n",
    "class Test_SW(unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a Classe SW\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    def test_rebuild (self):\n",
    "        \"\"\"Testar a função rebuild\n",
    "        \"\"\"\n",
    "\n",
    "        self.assertEqual(SW(\"ACTG\", \"ACTG\").rebuild(), ('ACTG', 'ACTG'))\n",
    "        self.assertEqual(SW(\"ATCATAA\", \"ATTACAT\").rebuild(), ('T-CAT', 'TACAT'))\n",
    "\n",
    "\n",
    "    def test_max_score (self):\n",
    "        \"\"\"Testar a funçao max_score\n",
    "        \"\"\"\n",
    "\n",
    "        self.assertEqual(SW(\"ACTG\", \"ACTG\").max_score(), 24)\n",
    "        self.assertEqual(SW(\"ATCATAA\", \"ATTACAT\").max_score(), 19)\n",
    "\n",
    "        \n",
    "unittest.main(argv=[''], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def consensus(s1, s2):\n",
    "    \"\"\"Função que permite obter o Consensus das 2 Sequências \n",
    "\n",
    "    Args:\n",
    "        s1 (str): Sequencia 1\n",
    "        s2 (str): Sequencia 2\n",
    "\n",
    "    Returns:\n",
    "        str: Consensus das Sequências\n",
    "    \"\"\"\n",
    "    res = \"\"\n",
    "    for x, y in zip(s1, s2):\n",
    "        if x == y:\n",
    "            res += x\n",
    "        elif x == '-':\n",
    "            res += y\n",
    "        else:\n",
    "            res += x\n",
    "    return res\n",
    "    \n",
    "class MultipleAlign:\n",
    "    \"\"\"Classe que permite realizar o alinhamento múltiplo de várias sequências \n",
    "    \"\"\"\n",
    "    def __init__(self, lista_seqs):\n",
    "        \"\"\"Contrutor da classe MultpleAlign\n",
    "\n",
    "        Args:\n",
    "            lista_seqs (list): Lista de sequências \n",
    "        \"\"\"\n",
    "        self.lista_seqs = lista_seqs\n",
    "        self.len = len(lista_seqs)\n",
    "\n",
    "\n",
    "\n",
    "    def __len__ (self):\n",
    "        \"\"\"Função que permite implementar a função len()\n",
    "\n",
    "        Returns:\n",
    "            int: Numero de Sequências presentes na Classe\n",
    "        \"\"\"\n",
    "\n",
    "        return len(self.lista_seqs)\n",
    "    \n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\"Função que permite representar a Classe como String \n",
    "\n",
    "        Returns:\n",
    "            str: Diferentes Sequências presentes na Classe\n",
    "        \"\"\"\n",
    "\n",
    "        return str(self.lista_seqs)\n",
    "    \n",
    "    def __getitem__(self,a):\n",
    "        \"\"\"Função que permite implementar a Indexação na Classe\n",
    "\n",
    "        Args:\n",
    "            a (int): Index\n",
    "\n",
    "        Returns:\n",
    "            str: Sequência correspondente ao Index \n",
    "        \"\"\"\n",
    "        return self.lista_seqs[a]\n",
    "    \n",
    "    \n",
    "    def align (self):\n",
    "        \"\"\"Função que permite realizar o Alinhamento Progressivo das Sequências \n",
    "\n",
    "        Returns:\n",
    "            list: Lista com o alinhamento entre todas as sequências\n",
    "        \"\"\"\n",
    "        lista_final_con = self.lista_seqs.copy()\n",
    "        lista_final = []\n",
    "        c = 0\n",
    "        while c < len(self.lista_seqs)-1:\n",
    "            lista_temp = lista_final_con[0:2].copy()\n",
    "            a = NW(lista_temp[0],lista_temp[1]).rebuild()\n",
    "            con = consensus(a[0],a[1])\n",
    "            del lista_final_con[0:2]\n",
    "            lista_final_con.insert(0,con)\n",
    "            c += 1\n",
    "\n",
    "        i = len(self.lista_seqs)\n",
    "\n",
    "        for c in range(i):\n",
    "            lista_final.append(NW(self.lista_seqs[c],lista_final_con[0]).rebuild()[0])\n",
    "        \n",
    "\n",
    "        return lista_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      ".....\n",
      "----------------------------------------------------------------------\n",
      "Ran 5 tests in 0.020s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b60493f40>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Test_MultipleAlign (unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a Classe MultipleAlign\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def test_align (self):\n",
    "        \"\"\"Testar função Align\n",
    "        \"\"\"\n",
    "        self.assertEqual(MultipleAlign([\"ATAGC\", \"AACC\", \"ATGAC\",\"ATAGC\", \"AACC\", \"ATGAC\"]).align(), ['ATAGC', 'A-ACC', 'ATGAC', 'ATAGC', 'A-ACC', 'ATGAC'])\n",
    "        self.assertEqual(MultipleAlign([\"TACCCGCT\",\"AACCA\"]).align(), ['TACCCGCT', '-A-AC-CA'])\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector as SQLC  \n",
    "\n",
    "DataBase = SQLC.connect(\n",
    "   host =\"127.0.0.1\",\n",
    "   user =\"root\",\n",
    "   password =\"login123\",\n",
    "   database = \"Base Dados IAP\",\n",
    "   auth_plugin=\"mysql_native_password\"\n",
    ")\n",
    "DataBase.autocommit= True\n",
    "cur = DataBase.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def connect_database (host =\"127.0.0.1\", user =\"root\",password =\"login123\", database = \"Base Dados IAP\"):\n",
    "    \"\"\"Funçao que permite ao utilizador conectar à base de dados (Base Dados IAP)\n",
    "\n",
    "    Args:\n",
    "        host (str, optional):  Host . Defaults to \"127.0.0.1\".\n",
    "        user (str, optional):  User . Defaults to \"root\".\n",
    "        password (str, optional):  Password . Defaults to \"login123\".\n",
    "        database (str, optional):  Database . Defaults to \"Base Dados IAP\".\n",
    "    \"\"\"\n",
    "\n",
    "    DataBase = SQLC.connect(\n",
    "\n",
    "        host = host,\n",
    "        user = user,\n",
    "        password = password,\n",
    "        database = database,\n",
    "        auth_plugin=\"mysql_native_password\"\n",
    "    )\n",
    "    DataBase.autocommit= True\n",
    "    cur = DataBase.cursor()\n",
    "\n",
    "def get_all_organisms ():\n",
    "    \"\"\"Função que permite obter todos os organismos presentes na Base de Dados Predefinida\n",
    "\n",
    "    Returns:\n",
    "        lista: Contêm nome de Organimos \n",
    "    \"\"\"\n",
    "    connect_database()\n",
    "    cur.execute(\"\"\"select Organism from Genbank\"\"\")\n",
    "    organisms_DB = cur.fetchall()\n",
    "\n",
    "    return organisms_DB\n",
    "\n",
    "\n",
    "def get_seq_from_organism (organism):\n",
    "    \"\"\"Função que permite obter a sequência do Organismo pretendido\n",
    "\n",
    "    Args:\n",
    "        organism (str): Organismo Presente na base de dados \n",
    "\n",
    "    Returns:\n",
    "        str: Sequência do Organismo\n",
    "    \"\"\"\n",
    "    connect_database()\n",
    "    cur.execute(f\"\"\" select Sequence from Sequences\n",
    "                    join Genbank on Sequences.ID_version_seq = Genbank.ID_version_seq\n",
    "                    where Organism = '{organism}'\n",
    "    \"\"\")\n",
    "\n",
    "    sequence = cur.fetchall()\n",
    "    \n",
    "    return sequence[0][0]\n",
    "\n",
    "def get_all_seqs():\n",
    "    \"\"\"Função que permite obter todas as sequências presentes na Base de Dados\n",
    "\n",
    "    Returns:\n",
    "        list: Lista com todas sequências presentes na Base de Dados\n",
    "    \"\"\"\n",
    "    connect_database()\n",
    "    cur.execute(f\"\"\" select Sequence from Sequences\n",
    "    \"\"\")\n",
    "\n",
    "    sequence = cur.fetchall()\n",
    "\n",
    "    return sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      ".......\n",
      "----------------------------------------------------------------------\n",
      "Ran 7 tests in 0.047s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b603324c0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "\n",
    "class Test_DataBase (unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a DataBase\n",
    "    \"\"\"\n",
    "\n",
    "    def test_database (self):\n",
    "        \"\"\"Testar a integridade da Base de Dados\n",
    "        \"\"\"\n",
    "        self.assertEqual(len(get_all_organisms()),len(get_all_seqs()))\n",
    "\n",
    "\n",
    "    def test_database_2 (self):\n",
    "        \"\"\"Testar a integridade da Base de Dados\n",
    "        \"\"\"\n",
    "        self.assertEqual(type(get_seq_from_organism('Sulfolobus spindle-shaped virus 1')), str)\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Blast ():\n",
    "    \"\"\"Classe que permite realizar um Blast Simplificado \n",
    "    \"\"\"\n",
    "    def __init__(self, query, seq, w = 3):\n",
    "        \"\"\"Construtor do Blast Simplificado \n",
    "\n",
    "        Args:\n",
    "            query (str): Query \n",
    "            seq (str): Seq\n",
    "            w (int, optional): W . Defaults to 3.\n",
    "            \n",
    "        \"\"\"\n",
    "        self.seq = seq\n",
    "        self.query = query\n",
    "        self.w = w\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.query)\n",
    "\n",
    "    \n",
    "    def query_map(query, w):\n",
    "        \"\"\"Função que permite devolver um dicionário em que as chaves são as sequências e os valores são uma lista dos índices\n",
    "\n",
    "        Args:\n",
    "            query (str): Query\n",
    "            w (int): W\n",
    "\n",
    "        Returns:\n",
    "            dict: Dicionário em que as chaves são as sequências e os valores são uma lista dos índice\n",
    "        \"\"\"\n",
    "        \n",
    "        tam = len(query)\n",
    "        res = {}\n",
    "        for chave, offset in [(query[p : p + w], p) for p in range(0, tam - w +1)]:\n",
    "            if chave not in res:\n",
    "                res[chave] = []\n",
    "            res[chave].append(offset)\n",
    "        return(res)\n",
    "\n",
    "\n",
    "    def get_all_offsets(s1, s2): \n",
    "        \"\"\"Função Auxiliar que permite devolver todos os índices das ocorrências de uma substring numa string\n",
    "\n",
    "        Args:\n",
    "            s1 (str): Substring\n",
    "            s2 (str): String\n",
    "\n",
    "        Returns:\n",
    "            dict: Dicionário com todas as ocorrèncias da substring na string\n",
    "        \"\"\"\n",
    "        w = len(s1)\n",
    "\n",
    "        return[p for p in range(0, len(s2) - w + 1) if s2[p : p + w] == s1]\n",
    "\n",
    "\n",
    "    def hits(qm, seq):\n",
    "        \"\"\"Função que permite gerar uma lista de hits\n",
    "\n",
    "        Args:\n",
    "            qm (dict): Dicionário proveninete da Função query_map\n",
    "            seq (str): Sequência Proveniente da base de dados\n",
    "        \n",
    "        Returns:\n",
    "            list: Lista de hits em que cada elemento é um tuplo com os indices\n",
    "        \"\"\"\n",
    "              \n",
    "        return[(o_query, o_seq) for chave, offsets in qm.items() \n",
    "        for o_query in offsets \n",
    "        for o_seq in Blast.get_all_offsets(chave, seq)]\n",
    "\n",
    "\n",
    "    def extend_hit_any_direction(query, seq, o1, o2, dir):\n",
    "        \"\"\"Função auxliar que permite extender o hit em ambas as direções\n",
    "\n",
    "        \"\"\"\n",
    "        matches = 0\n",
    "        count = 0\n",
    "        while o1 >= 0 and o2 >= 0 and o1 < len(query) and o2 < len(seq):\n",
    "            matches += 1 if query[o1] == seq[o2] else 0\n",
    "            count += 1\n",
    "            if 2 * matches < count:\n",
    "                return o1, o2, matches, count\n",
    "            o1 += dir\n",
    "            o2 += dir\n",
    "        return o1 - dir, o2 - dir, matches, count\n",
    "\n",
    "\n",
    "    def extend_hit(query, seq, hit, w):\n",
    "        \"\"\"Função que permite estende um hit em cada direção se o nº de matches for de pelo menos metade do tamanho da extensão\n",
    "\n",
    "        Args:\n",
    "            query (str): Query\n",
    "            seq (str): Seq\n",
    "            hit (tuple): Hit\n",
    "            w (int): w\n",
    "\n",
    "        Returns:\n",
    "            tuple: Tuplo com o índice do início do hit estendido na query, na sequência, o tamanho e o nºde matches\n",
    "        \"\"\"\n",
    "\n",
    "        o1, o2 = hit\n",
    "        left  = Blast.extend_hit_any_direction(query, seq, o1 - 1, o2 - 1, -1)\n",
    "        right = Blast.extend_hit_any_direction(query, seq, o1 + w, o2 + w, +1)\n",
    "\n",
    "        O1, O2, ML, SL = left\n",
    "        _,   _, MR, SR = right\n",
    "        return O1, O2, w + SL + SR, ML + w + MR\n",
    "    \n",
    "\n",
    "    def best_hit (self):\n",
    "        \"\"\"Função que permite devolver a extensão de maior score (no caso de empate, deverá devolver a de menor tamanho que aparece primeiro)\n",
    "\n",
    "        Returns:\n",
    "            tuple: Tuplo com a extensão que possui maior score\n",
    "        \"\"\"\n",
    "        qm = Blast.query_map(self.query,self.w)\n",
    "        lista_hits = Blast.hits(qm, self.seq)\n",
    "        all_scores =[Blast.extend_hit(self.query, self.seq, c, self.w) for c in lista_hits]\n",
    "        all_scores.sort(key=lambda c: c[2] and c[3], reverse=True)\n",
    "        if all_scores == []:\n",
    "            return print(\"Not Found\")\n",
    "        else:\n",
    "            return all_scores [0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "..............\n",
      "----------------------------------------------------------------------\n",
      "Ran 14 tests in 0.058s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b60342880>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "\n",
    "query = \"AATATAT\"\n",
    "seq = \"AATATGTTATATAATAATATTT\"\n",
    "w = 3\n",
    "\n",
    "class Test_Blast (unittest.TestCase):\n",
    "    \"\"\"Estes Testes foram baseados nos exemplos fornecidos pelo Professor Rui (inputs/outputs) durantes\n",
    "    as aulas teóricas \n",
    "    \"\"\"\n",
    "\n",
    "    def test_query_map (self):\n",
    "        \"\"\"Testar a função query_map\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.query_map(query,w),{'AAT': [0], 'ATA': [1, 3], 'TAT': [2, 4]})\n",
    "    \n",
    "\n",
    "    def test_get_all_offsets (self):\n",
    "        \"\"\"Testar a função get_all_offsets\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.get_all_offsets(query,seq), [])\n",
    "    \n",
    "    \n",
    "    def test_hits (self):\n",
    "        \"\"\"Testar a função test_hits\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.hits(Blast.query_map(query,w),seq),[\n",
    "            (0, 0), (0, 12), (0, 15), (1, 1), (1, 8),\n",
    "            (1, 10), (1, 13), (1, 16), (3, 1), (3, 8), (3, 10),\n",
    "            (3, 13), (3, 16), (2, 2), (2, 7), (2, 9), (2, 17),\n",
    "            (4, 2), (4, 7), (4, 9),(4, 17)])\n",
    "\n",
    "\n",
    "    def test_extend_hit_any_direction_left (self):\n",
    "        \"\"\"Testar a função extend_hit_any_direction\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.extend_hit_any_direction(query, seq, 1, 16, -1),(0, 15, 2, 2))\n",
    "\n",
    "    \n",
    "    def test_extend_hit_any_direction_right (self):\n",
    "        \"\"\"Testar a função extend_hit_any_direction\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.extend_hit_any_direction(query, seq, 1, 16, 1),(6, 21, 5, 6))\n",
    "\n",
    "    def test_extend_hit (self):\n",
    "        \"\"\"Testar a função extend_hit\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast.extend_hit(query, seq, (1, 16), w), (0, 15, 7, 6))\n",
    "\n",
    "    \n",
    "    def test_best_hit (self):\n",
    "        \"\"\"Testar a função best_hit\n",
    "        \"\"\"\n",
    "        self.assertEqual(Blast(query, seq, w).best_hit(), (0, 0, 7, 6))\n",
    "\n",
    "unittest.main(argv=[''], exit=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Luis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Motifs:\n",
    "    \"\"\"Classe que permite gerar Motifs probabilisticos \n",
    "    \"\"\"\n",
    "    def __init__(self, list_seqs):\n",
    "        \"\"\"Contrutor do Motifs onde criámos o nosso Alfabeto(letters) e onde gerámos a nossa Matriz vazia (mat)\n",
    "\n",
    "        Args:\n",
    "            list_seqs (list): Lista contendo diferentes sequências biológicas\n",
    "        \"\"\"\n",
    "        from collections import Counter\n",
    "        self.seqs = list_seqs\n",
    "        self.letters = list(Counter([a for c in list_seqs for a in c]))\n",
    "        self.mat = Mat(len(self.letters)+1,len(list_seqs[0])+1)\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return print(self.mat)\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        return print(self.mat)\n",
    "\n",
    "    def contagens (self):\n",
    "        \"\"\"Função que permite devolver a frequência dos elementos das sequências\n",
    "\n",
    "        Returns:\n",
    "            list: Matriz das Contagens \n",
    "        \"\"\"\n",
    "        for c in self.seqs:\n",
    "            for coluna in range(len(self.seqs[0])):\n",
    "                linha = self.letters.index(c[coluna])\n",
    "                self.mat[linha + 1][coluna + 1] += 1\n",
    "                self.mat[0][coluna + 1] = coluna +1 \n",
    "                self.mat[0][0] = \".\"                \n",
    "                self.mat[linha +1][0] = self.letters[linha] \n",
    "        return print(self.mat)\n",
    "\n",
    "\n",
    "    def PWM (self, pseudo_count = 0):\n",
    "        \"\"\"Função que permite gerar a Matriz PWM\n",
    "\n",
    "        Args:\n",
    "            pseudo_count (int, optional): pseudocontagem. Defaults to 0.\n",
    "\n",
    "        Returns:\n",
    "            list: Matriz PWM\n",
    "        \"\"\"\n",
    "        self.mat = Mat(len(self.letters)+1,len(self.seqs[0])+1)\n",
    "        for c in self.seqs:\n",
    "            for coluna in range(len(self.seqs[0])):\n",
    "                linha = self.letters.index(c[coluna])\n",
    "                self.mat[linha + 1][coluna + 1] += 1\n",
    "                self.mat[0][coluna + 1] = f\"  {coluna +1} \"   \" \"  \n",
    "                self.mat[0][0] = \".\"                \n",
    "                self.mat[linha +1][0] = self.letters[linha] \n",
    "        try:      \n",
    "            for c in range(1,len(self.letters)+ 1):\n",
    "                for a in range(1,len(self.seqs)+ 1): \n",
    "                    self.mat[c][a] = round(float(self.mat[c][a] + pseudo_count + 0.01) / float(len(self.seqs)),3)\n",
    "            return print(self.mat)\n",
    "\n",
    "        except:\n",
    "            for c in range(1,len(self.letters)+ 1):\n",
    "                for a in range(1,len(self.seqs)+ 2): \n",
    "                    self.mat[c][a] = round(float(self.mat[c][a] + pseudo_count + 0.01) / float(len(self.seqs)),3)\n",
    "            return print(self.mat)\n",
    "\n",
    "    \n",
    "    def PSSM (self, pseudo_count = 0):\n",
    "        \"\"\"Função que permite gerar a Matriz PSSM\n",
    "\n",
    "        Args:\n",
    "            pseudo_count (int, optional): pseudocontagem. Defaults to 0.\n",
    "\n",
    "        Returns:\n",
    "            list: Matriz PSSM\n",
    "        \"\"\"\n",
    "        self.mat = Mat(len(self.letters)+1,len(self.seqs[0])+1)\n",
    "        from math import log2\n",
    "        for c in self.seqs:\n",
    "            for coluna in range(len(self.seqs[0])):\n",
    "                linha = self.letters.index(c[coluna])\n",
    "                self.mat[linha + 1][coluna + 1] += 1\n",
    "                self.mat[0][coluna + 1] = f\"  {coluna +1} \"   \" \"  \n",
    "                self.mat[0][0] = \".\"                \n",
    "                self.mat[linha +1][0] = self.letters[linha] \n",
    "        try:\n",
    "            for c in range(1,len(self.letters)+ 1):\n",
    "                for a in range(1,len(self.seqs)+ 1): \n",
    "                    self.mat[c][a] = round((log2(((self.mat[c][a]+ pseudo_count)/(len(self.seqs)+ len(self.letters)* pseudo_count)/(1/len(self.letters)) ))),3)\n",
    "            return print(self.mat)\n",
    "        except:\n",
    "            for c in range(1,len(self.letters)+ 1):\n",
    "                for a in range(1,len(self.seqs)+ 2): \n",
    "                    self.mat[c][a] = round((log2(((self.mat[c][a]+ pseudo_count)/(len(self.seqs)+ len(self.letters* pseudo_count)/(1/len(self.letters)) )))),3)\n",
    "            return print(self.mat)\n",
    "    \n",
    "    def prob_seq(self, seq):\n",
    "        \"\"\"Função que permite devolver a probabilidade de uma sequência\n",
    "\n",
    "        Args:\n",
    "            seq (str): Sequência \n",
    "\n",
    "        Returns:\n",
    "            float: Probabilidade da sequência \n",
    "        \"\"\"\n",
    "\n",
    "        self.mat = Mat(len(self.letters)+1,len(self.seqs[0])+1)\n",
    "        prob = 1 \n",
    "        for c in self.seqs:\n",
    "            for coluna in range(len(self.seqs[0])):\n",
    "                linha = self.letters.index(c[coluna])\n",
    "                self.mat[linha + 1][coluna + 1] += 1\n",
    "                self.mat[0][coluna + 1] = f\"  {coluna +1} \"   \" \"  \n",
    "                self.mat[0][0] = \".\"                \n",
    "                self.mat[linha +1][0] = self.letters[linha] \n",
    "        for c in range(1,len(self.letters)+ 1):\n",
    "            for a in range(1,len(self.seqs)+ 1):\n",
    "                self.mat[c][a] = round(float(self.mat[c][a] + 0.01) / float(len(self.seqs)),3)\n",
    "        \n",
    "        for c in range(1,len(self.seqs[0])+ 1):\n",
    "            a = self.letters.index(seq[c - 1])\n",
    "            prob *= self.mat[a +1][c]\n",
    "        return round(prob,4)\n",
    "\n",
    "\n",
    "    def most_prob_seq (self,pseudo_count = 0):\n",
    "        \"\"\"Função que permite devolver a sequência mais provável de ocorrer\n",
    "\n",
    "        Args:\n",
    "            pseudo_count (int, optional): pseudocontagem. Defaults to 0.\n",
    "\n",
    "        Returns:\n",
    "            str: Sequência mais provável de ocorrer\n",
    "        \"\"\"\n",
    "\n",
    "        self.mat = Mat(len(self.letters)+1,len(self.seqs[0])+1)\n",
    "        resul = \"\"\n",
    "        for c in self.seqs:\n",
    "            for coluna in range(len(self.seqs[0])):\n",
    "                linha = self.letters.index(c[coluna])\n",
    "                self.mat[linha + 1][coluna + 1] += 1\n",
    "                self.mat[0][coluna + 1] = f\"  {coluna +1} \"   \" \"  \n",
    "                self.mat[0][0] = \".\"               \n",
    "                self.mat[linha +1][0] = self.letters[linha] \n",
    "        for c in range(1,len(self.letters)+ 1):\n",
    "            for a in range(1,len(self.seqs)+ 1):\n",
    "                self.mat[c][a] = round(float(self.mat[c][a] + pseudo_count + 0.01) / float(len(self.seqs)),3)\n",
    "        list_temp= []\n",
    "        list_final = []\n",
    "        for c in range(1,len(self.seqs)+1):\n",
    "            for a in range(1,len(self.seqs)+ 1):\n",
    "                list_temp.append((self.mat[a][c],a))\n",
    "            list_final.append(max(list_temp))\n",
    "            list_temp =[]\n",
    "        \n",
    "        for c in list_final:\n",
    "            for a in range(1,len(self.seqs[0])+1):\n",
    "                if a == c[1]:\n",
    "                    resul += self.mat[a][0]\n",
    "        return resul\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "................\n",
      "----------------------------------------------------------------------\n",
      "Ran 16 tests in 0.052s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.main.TestProgram at 0x7f1b6033f250>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest  \n",
    "seqs = ['ATTG','ATCG','ATTC','ACTC']\n",
    "\n",
    "class Test_Motifs (unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a Classe Motifs\n",
    "    \"\"\"\n",
    "\n",
    "    def test_prob_seq (self):\n",
    "        \"\"\"Testar a função prob_seq\n",
    "        \"\"\"\n",
    "\n",
    "        self.assertEqual(Motifs(seqs).prob_seq(\"ATTG\"), 0.2845)\n",
    "\n",
    "\n",
    "    def test_most_prob_seq (self):\n",
    "        \"\"\"Testar a função most_prob_seq\n",
    "        \"\"\"\n",
    "        \n",
    "        self.assertEqual(Motifs(seqs).most_prob_seq(), \"ATTC\")\n",
    "\n",
    "unittest.main(argv=[''], exit=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Enzyme_Restriction:\n",
    "    def __init__(self, enzyme, seq):\n",
    "        \"\"\"Construtor da classe Enzyme_Restriction.\n",
    "\n",
    "        Args:\n",
    "            enzyme (str): Sequência da enzima\n",
    "            seq (str): Sequência nucleótida\n",
    "        \"\"\"\n",
    "        self.enzyme_seq = enzyme \n",
    "        self.seq = seq \n",
    "\n",
    "    def __str__ (self):\n",
    "        \"\"\"Função que permite representar as sequências como uma string.\n",
    "\n",
    "        Returns:\n",
    "            str: Representação das sequências\n",
    "        \"\"\"\n",
    "        return \"Enzyme sequence = \" + str(self.enzyme_seq) + \"\\nNucleotide sequence = \" + str(self.seq)\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\"Função que permite representar as sequências como uma string na shell.\n",
    "\n",
    "        Returns:\n",
    "            str: Representação das sequências\n",
    "        \"\"\"\n",
    "        return f\"Enzyme_Restriction(\\\"{self.enzyme_seq}\\\",\\\"{self.seq}\\\")\"\n",
    "\n",
    "    def IUB(self):\n",
    "        \"\"\"Conversão da enzima de restrição na expressao regular correspondente.\n",
    "\n",
    "        Returns:\n",
    "            str: Expressão regular\n",
    "        \"\"\"\n",
    "        try:\n",
    "            rebase={\"R\":\"[GA]\", \"Y\":\"[CT]\",\"M\":\"[AC]\", \"K\":\"[GT]\", \"S\":\"[GC]\", \"W\": \"[AT]\", \"B\":\"[CTG]\", \"D\":\"[ATG]\", \"H\":\"[ACT]\",\"V\":\"[ACG]\", \"N\":\"[ACGT]\", \"A\":\"A\",\"C\":\"C\",\"G\":\"G\",\"T\":\"T\"}\n",
    "            cut = self.enzyme_seq.replace(\"^\",\"\")\n",
    "            regexp = \"\"\n",
    "            for x in cut:\n",
    "                regexp += rebase[x]\n",
    "            return regexp\n",
    "        except:\n",
    "            return \"Nucleotide sequence or Enzyme invalid\"\n",
    "\n",
    "    def cut_positions(self): \n",
    "        \"\"\"Função que permite determinar as posições onde a enzima de restrição \"corta\" a sequência.\n",
    "\n",
    "        Returns:\n",
    "            list: lista das posições de \"corte\"\n",
    "        \"\"\"\n",
    "        from re import finditer\n",
    "        cutposition = self.enzyme_seq.find(\"^\")\n",
    "        regexp = Enzyme_Restriction.IUB(self)\n",
    "        matches = finditer(regexp, self.seq)\n",
    "        locs = []\n",
    "        for x in matches:\n",
    "            locs.append(x.start() + cutposition)\n",
    "        return locs\n",
    "\n",
    "    def cut_positions_re(self): \n",
    "        \"\"\"Função que permite determinar as posições onde a enzima de restrição \"corta\" a sequência complementar.\n",
    "\n",
    "        Returns:\n",
    "            list: lista das posições de \"corte\"\n",
    "        \"\"\"\n",
    "        from re import finditer\n",
    "        cutposition = self.enzyme_seq.find(\"^\")\n",
    "        seq_trans = self.seq.replace(\"A\",\"t\").replace(\"T\",\"a\").replace(\"C\",\"g\").replace(\"G\",\"c\")[::-1].upper()\n",
    "        regexp = Enzyme_Restriction.IUB(self)\n",
    "        matches = finditer(regexp, seq_trans)\n",
    "        locs = []\n",
    "        for x in matches:\n",
    "            locs.append(x.start() + cutposition)\n",
    "        return locs\n",
    "\n",
    "    def cut_subsequences(self):\n",
    "        \"\"\"Função que determina as subsequências resultantes do corte da sequência (seq) usando os índices na lista locs.\n",
    "\n",
    "        Returns:\n",
    "            list: lista das subsequências\n",
    "        \"\"\"\n",
    "        res =[]\n",
    "        positions = Enzyme_Restriction.cut_positions(self)\n",
    "        positions.insert(0,0)\n",
    "        positions.append(len(self.seq))\n",
    "        for i in range (len(positions)-1):\n",
    "            res.append(self.seq[positions[i]:positions[i+1]])\n",
    "        return res\n",
    "\n",
    "    def cut_subsequences_re(self):\n",
    "        \"\"\"Função que determina as subsequências resultantes do corte da sequência completamentar (seq) usando os índices na lista locs.\n",
    "\n",
    "        Returns:\n",
    "            list: lista das subsequências\n",
    "        \"\"\"\n",
    "        res =[]\n",
    "        positions = Enzyme_Restriction(self.enzyme_seq,self.seq).cut_positions()\n",
    "        positions.insert(0,0)\n",
    "        seq_trans = self.seq.replace(\"A\",\"t\").replace(\"T\",\"a\").replace(\"C\",\"g\").replace(\"G\",\"c\")[::-1].upper()\n",
    "        positions.append(len(self.seq))\n",
    "        for i in range (len(positions)-1):\n",
    "            res.append(seq_trans[positions[i]:positions[i+1]])\n",
    "        return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "class Test_Enzyme_Restriction(unittest.TestCase):\n",
    "\n",
    "    def test_IUB(self):\n",
    "        \"\"\"Função usada para testar a função IUB() da classe Enzyme_Restriction.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"TAAGACTGAAT\").IUB(),\"GA[AC]T[ACG]\")\n",
    "        self.assertEqual(Enzyme_Restriction(\"GPAMTV\",\"TAAGACTGAAT\").IUB(),\"Nucleotide sequence or Enzyme invalid\")\n",
    "        self.assertEqual(Enzyme_Restriction(\"GPAMTV\",\"TGCAACGU\").IUB(),\"Nucleotide sequence or Enzyme invalid\")\n",
    "        self.assertEqual(Enzyme_Restriction(\"G_AMTV\",\"TAAGACTGAAT\").IUB(),\"Nucleotide sequence or Enzyme invalid\")\n",
    "    \n",
    "    def test_cut_positions(self):\n",
    "        \"\"\"Função usada para testar a função cut_positions() da classe Enzyme_Restriction.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"TAAGACTGAAT\").cut_positions(), [4])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GAAACT\",\"TAAGACTGAAT\").cut_positions(), [])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA    AACT\",\"TAAGACTGAAT\").cut_positions(), [])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA*_*PAACT\",\"TAAGACTGAAT\").cut_positions(), [])\n",
    "\n",
    "    def test_cut_positions_re(self):\n",
    "        \"\"\"Função usada para testar a função cut_positions_re() da classe Enzyme_Restriction.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"ATTCAGTCTTA\").cut_positions_re(), [4])\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"ATAGTCTCAGTCTTA\").cut_positions_re(), [4, 10])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA    AACT\",\"ATAGTCTCAGTCTTA\").cut_positions_re(), [])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA12AACT\",\"TAAGACTGAAT\").cut_positions_re(), [])\n",
    "\n",
    "    def test_cut_subsequences(self):\n",
    "        \"\"\"Função usada para testar a função cut_subsequences() da classe Enzyme_Restriction.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"TAAGACTGAAT\").cut_subsequences(), ['TAAG', 'ACTGAAT'])\n",
    "        self.assertEqual(Enzyme_Restriction('GA*_*PAACT', 'TAAGACTGAAT').cut_subsequences(), [\"TAAGACTGAAT\"])\n",
    "        self.assertEqual(Enzyme_Restriction('GA   PAACT', 'TAAGACTGAAT').cut_subsequences(), [\"TAAGACTGAAT\"])\n",
    "        self.assertEqual(Enzyme_Restriction('PPPAA12', 'TAAGACTGAAT').cut_subsequences(), [\"TAAGACTGAAT\"])\n",
    "\n",
    "    def test_cut_subsequences_re(self):\n",
    "        \"\"\"Função usada para testar a função cut_subsequences_re() da classe Enzyme_Restriction.\n",
    "        \"\"\"\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"GTAGAAGATTCT\").cut_subsequences_re(), ['AGAATCTTCTAC'])\n",
    "        self.assertEqual(Enzyme_Restriction(\"G^AMTV\",\"TAAGACTGAAT\").cut_subsequences_re(), ['ATTC', 'AGTCTTA'])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA   PAACT\",\"ATTCAGTCTTA\").cut_subsequences_re(), ['TAAGACTGAAT'])\n",
    "        self.assertEqual(Enzyme_Restriction(\"GA1223PAACT\",\"ATTCAGTCTTA\").cut_subsequences_re(), ['TAAGACTGAAT'])\n",
    "\n",
    "\n",
    "unittest.main(argv=[\"\"], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prosite(seq, pattern): \n",
    "    \"\"\"Função que permite converter uma expressão do padrão Prosite numa expressão regular\n",
    "\n",
    "    Returns:\n",
    "        str: Expressão regular\n",
    "    \"\"\"\n",
    "    from re import search\n",
    "    er = pattern.replace(\"-\",\"\")\n",
    "    er = er.replace(\"x\",\".\")\n",
    "    er = er.replace(\"(\",\"{\")\n",
    "    er = er.replace(\")\",\"}\")\n",
    "    motif = search(er, seq)\n",
    "\n",
    "    if (motif != None):\n",
    "        zona = motif.span()              \n",
    "        return (\"Motifs: {}\" .format(seq[zona[0]:zona[1]]))\n",
    "    else :\n",
    "        return \"Not found\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "\n",
    "class Test_Prosite(unittest.TestCase):\n",
    "    \"\"\"Conjunto de testes para a classe Prosite\n",
    "    \"\"\"\n",
    "\n",
    "    def test_prosite(self):\n",
    "        \"\"\"Testar a função prosite\n",
    "        \"\"\"\n",
    "        self.assertEqual(prosite(\"HKMMLASCKHLLCLKCIVKLG\",\"C-x-H-x-[LIVMFY]-C-x(2)-C-[LIVMYA]\"),\"Motifs: CKHLLCLKCI\")\n",
    "        self.assertEqual(prosite(\"HKMMLLCLKCIVKLG\",\"C-x-H-x-[LIVMFY]-C-x(2)-C-[LIVMYA]\"),\"Not found\")\n",
    "\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mestrado_final",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 (main, Oct 13 2022, 21:15:33) \n[GCC 11.2.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f2307cf6e279acbb4fe5a904c1ba66075afa04f4e5f00d0d9abed1d852b5aca9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
